// weibo/lib-cov/oauth.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['oauth.js']) {
  _$jscoverage['oauth.js'] = [];
  _$jscoverage['oauth.js'][42] = 0;
  _$jscoverage['oauth.js'][44] = 0;
  _$jscoverage['oauth.js'][45] = 0;
  _$jscoverage['oauth.js'][46] = 0;
  _$jscoverage['oauth.js'][48] = 0;
  _$jscoverage['oauth.js'][51] = 0;
  _$jscoverage['oauth.js'][53] = 0;
  _$jscoverage['oauth.js'][54] = 0;
  _$jscoverage['oauth.js'][55] = 0;
  _$jscoverage['oauth.js'][56] = 0;
  _$jscoverage['oauth.js'][59] = 0;
  _$jscoverage['oauth.js'][63] = 0;
  _$jscoverage['oauth.js'][65] = 0;
  _$jscoverage['oauth.js'][66] = 0;
  _$jscoverage['oauth.js'][68] = 0;
  _$jscoverage['oauth.js'][69] = 0;
  _$jscoverage['oauth.js'][70] = 0;
  _$jscoverage['oauth.js'][71] = 0;
  _$jscoverage['oauth.js'][72] = 0;
  _$jscoverage['oauth.js'][74] = 0;
  _$jscoverage['oauth.js'][76] = 0;
  _$jscoverage['oauth.js'][78] = 0;
  _$jscoverage['oauth.js'][83] = 0;
  _$jscoverage['oauth.js'][84] = 0;
  _$jscoverage['oauth.js'][85] = 0;
  _$jscoverage['oauth.js'][87] = 0;
  _$jscoverage['oauth.js'][88] = 0;
  _$jscoverage['oauth.js'][89] = 0;
  _$jscoverage['oauth.js'][94] = 0;
  _$jscoverage['oauth.js'][95] = 0;
  _$jscoverage['oauth.js'][97] = 0;
  _$jscoverage['oauth.js'][98] = 0;
  _$jscoverage['oauth.js'][100] = 0;
  _$jscoverage['oauth.js'][101] = 0;
  _$jscoverage['oauth.js'][103] = 0;
  _$jscoverage['oauth.js'][104] = 0;
  _$jscoverage['oauth.js'][105] = 0;
  _$jscoverage['oauth.js'][107] = 0;
  _$jscoverage['oauth.js'][111] = 0;
  _$jscoverage['oauth.js'][112] = 0;
  _$jscoverage['oauth.js'][114] = 0;
  _$jscoverage['oauth.js'][115] = 0;
  _$jscoverage['oauth.js'][117] = 0;
  _$jscoverage['oauth.js'][118] = 0;
  _$jscoverage['oauth.js'][119] = 0;
  _$jscoverage['oauth.js'][120] = 0;
  _$jscoverage['oauth.js'][121] = 0;
  _$jscoverage['oauth.js'][122] = 0;
  _$jscoverage['oauth.js'][125] = 0;
  _$jscoverage['oauth.js'][127] = 0;
  _$jscoverage['oauth.js'][130] = 0;
  _$jscoverage['oauth.js'][131] = 0;
  _$jscoverage['oauth.js'][132] = 0;
  _$jscoverage['oauth.js'][133] = 0;
  _$jscoverage['oauth.js'][134] = 0;
  _$jscoverage['oauth.js'][135] = 0;
  _$jscoverage['oauth.js'][136] = 0;
  _$jscoverage['oauth.js'][138] = 0;
  _$jscoverage['oauth.js'][139] = 0;
  _$jscoverage['oauth.js'][141] = 0;
  _$jscoverage['oauth.js'][143] = 0;
  _$jscoverage['oauth.js'][146] = 0;
  _$jscoverage['oauth.js'][147] = 0;
  _$jscoverage['oauth.js'][148] = 0;
  _$jscoverage['oauth.js'][149] = 0;
  _$jscoverage['oauth.js'][150] = 0;
  _$jscoverage['oauth.js'][151] = 0;
  _$jscoverage['oauth.js'][153] = 0;
  _$jscoverage['oauth.js'][154] = 0;
  _$jscoverage['oauth.js'][155] = 0;
  _$jscoverage['oauth.js'][156] = 0;
  _$jscoverage['oauth.js'][157] = 0;
  _$jscoverage['oauth.js'][158] = 0;
  _$jscoverage['oauth.js'][160] = 0;
  _$jscoverage['oauth.js'][161] = 0;
  _$jscoverage['oauth.js'][163] = 0;
  _$jscoverage['oauth.js'][165] = 0;
  _$jscoverage['oauth.js'][168] = 0;
  _$jscoverage['oauth.js'][169] = 0;
  _$jscoverage['oauth.js'][170] = 0;
  _$jscoverage['oauth.js'][171] = 0;
  _$jscoverage['oauth.js'][172] = 0;
  _$jscoverage['oauth.js'][173] = 0;
  _$jscoverage['oauth.js'][175] = 0;
  _$jscoverage['oauth.js'][176] = 0;
  _$jscoverage['oauth.js'][180] = 0;
  _$jscoverage['oauth.js'][181] = 0;
  _$jscoverage['oauth.js'][184] = 0;
  _$jscoverage['oauth.js'][185] = 0;
  _$jscoverage['oauth.js'][186] = 0;
  _$jscoverage['oauth.js'][190] = 0;
  _$jscoverage['oauth.js'][191] = 0;
  _$jscoverage['oauth.js'][192] = 0;
  _$jscoverage['oauth.js'][196] = 0;
  _$jscoverage['oauth.js'][197] = 0;
  _$jscoverage['oauth.js'][200] = 0;
  _$jscoverage['oauth.js'][201] = 0;
  _$jscoverage['oauth.js'][202] = 0;
  _$jscoverage['oauth.js'][203] = 0;
  _$jscoverage['oauth.js'][204] = 0;
  _$jscoverage['oauth.js'][206] = 0;
  _$jscoverage['oauth.js'][208] = 0;
  _$jscoverage['oauth.js'][211] = 0;
  _$jscoverage['oauth.js'][215] = 0;
  _$jscoverage['oauth.js'][216] = 0;
  _$jscoverage['oauth.js'][217] = 0;
  _$jscoverage['oauth.js'][219] = 0;
  _$jscoverage['oauth.js'][220] = 0;
  _$jscoverage['oauth.js'][221] = 0;
  _$jscoverage['oauth.js'][222] = 0;
  _$jscoverage['oauth.js'][223] = 0;
  _$jscoverage['oauth.js'][224] = 0;
  _$jscoverage['oauth.js'][227] = 0;
  _$jscoverage['oauth.js'][231] = 0;
  _$jscoverage['oauth.js'][235] = 0;
  _$jscoverage['oauth.js'][236] = 0;
  _$jscoverage['oauth.js'][238] = 0;
  _$jscoverage['oauth.js'][239] = 0;
  _$jscoverage['oauth.js'][240] = 0;
  _$jscoverage['oauth.js'][241] = 0;
  _$jscoverage['oauth.js'][242] = 0;
  _$jscoverage['oauth.js'][244] = 0;
  _$jscoverage['oauth.js'][248] = 0;
  _$jscoverage['oauth.js'][254] = 0;
  _$jscoverage['oauth.js'][255] = 0;
  _$jscoverage['oauth.js'][256] = 0;
  _$jscoverage['oauth.js'][257] = 0;
  _$jscoverage['oauth.js'][258] = 0;
  _$jscoverage['oauth.js'][259] = 0;
  _$jscoverage['oauth.js'][260] = 0;
  _$jscoverage['oauth.js'][264] = 0;
  _$jscoverage['oauth.js'][268] = 0;
  _$jscoverage['oauth.js'][271] = 0;
  _$jscoverage['oauth.js'][274] = 0;
  _$jscoverage['oauth.js'][275] = 0;
  _$jscoverage['oauth.js'][277] = 0;
  _$jscoverage['oauth.js'][278] = 0;
  _$jscoverage['oauth.js'][282] = 0;
  _$jscoverage['oauth.js'][283] = 0;
  _$jscoverage['oauth.js'][284] = 0;
  _$jscoverage['oauth.js'][286] = 0;
  _$jscoverage['oauth.js'][288] = 0;
  _$jscoverage['oauth.js'][297] = 0;
  _$jscoverage['oauth.js'][299] = 0;
  _$jscoverage['oauth.js'][300] = 0;
  _$jscoverage['oauth.js'][301] = 0;
  _$jscoverage['oauth.js'][302] = 0;
  _$jscoverage['oauth.js'][304] = 0;
  _$jscoverage['oauth.js'][309] = 0;
  _$jscoverage['oauth.js'][310] = 0;
  _$jscoverage['oauth.js'][311] = 0;
  _$jscoverage['oauth.js'][312] = 0;
  _$jscoverage['oauth.js'][313] = 0;
  _$jscoverage['oauth.js'][315] = 0;
  _$jscoverage['oauth.js'][316] = 0;
  _$jscoverage['oauth.js'][317] = 0;
  _$jscoverage['oauth.js'][318] = 0;
  _$jscoverage['oauth.js'][319] = 0;
  _$jscoverage['oauth.js'][321] = 0;
  _$jscoverage['oauth.js'][323] = 0;
  _$jscoverage['oauth.js'][324] = 0;
  _$jscoverage['oauth.js'][334] = 0;
  _$jscoverage['oauth.js'][335] = 0;
  _$jscoverage['oauth.js'][340] = 0;
  _$jscoverage['oauth.js'][341] = 0;
  _$jscoverage['oauth.js'][342] = 0;
  _$jscoverage['oauth.js'][344] = 0;
  _$jscoverage['oauth.js'][347] = 0;
  _$jscoverage['oauth.js'][348] = 0;
  _$jscoverage['oauth.js'][349] = 0;
  _$jscoverage['oauth.js'][352] = 0;
  _$jscoverage['oauth.js'][353] = 0;
  _$jscoverage['oauth.js'][354] = 0;
  _$jscoverage['oauth.js'][355] = 0;
  _$jscoverage['oauth.js'][356] = 0;
  _$jscoverage['oauth.js'][359] = 0;
  _$jscoverage['oauth.js'][360] = 0;
  _$jscoverage['oauth.js'][361] = 0;
  _$jscoverage['oauth.js'][362] = 0;
  _$jscoverage['oauth.js'][365] = 0;
  _$jscoverage['oauth.js'][370] = 0;
  _$jscoverage['oauth.js'][371] = 0;
  _$jscoverage['oauth.js'][372] = 0;
  _$jscoverage['oauth.js'][373] = 0;
  _$jscoverage['oauth.js'][374] = 0;
  _$jscoverage['oauth.js'][376] = 0;
  _$jscoverage['oauth.js'][377] = 0;
  _$jscoverage['oauth.js'][378] = 0;
  _$jscoverage['oauth.js'][381] = 0;
  _$jscoverage['oauth.js'][386] = 0;
  _$jscoverage['oauth.js'][392] = 0;
  _$jscoverage['oauth.js'][395] = 0;
  _$jscoverage['oauth.js'][396] = 0;
  _$jscoverage['oauth.js'][397] = 0;
  _$jscoverage['oauth.js'][398] = 0;
  _$jscoverage['oauth.js'][399] = 0;
  _$jscoverage['oauth.js'][401] = 0;
  _$jscoverage['oauth.js'][404] = 0;
  _$jscoverage['oauth.js'][405] = 0;
  _$jscoverage['oauth.js'][407] = 0;
  _$jscoverage['oauth.js'][408] = 0;
  _$jscoverage['oauth.js'][409] = 0;
  _$jscoverage['oauth.js'][410] = 0;
  _$jscoverage['oauth.js'][411] = 0;
  _$jscoverage['oauth.js'][412] = 0;
  _$jscoverage['oauth.js'][415] = 0;
  _$jscoverage['oauth.js'][416] = 0;
  _$jscoverage['oauth.js'][417] = 0;
  _$jscoverage['oauth.js'][418] = 0;
  _$jscoverage['oauth.js'][419] = 0;
  _$jscoverage['oauth.js'][420] = 0;
  _$jscoverage['oauth.js'][422] = 0;
  _$jscoverage['oauth.js'][426] = 0;
  _$jscoverage['oauth.js'][429] = 0;
  _$jscoverage['oauth.js'][433] = 0;
  _$jscoverage['oauth.js'][436] = 0;
  _$jscoverage['oauth.js'][440] = 0;
  _$jscoverage['oauth.js'][441] = 0;
  _$jscoverage['oauth.js'][442] = 0;
  _$jscoverage['oauth.js'][443] = 0;
  _$jscoverage['oauth.js'][445] = 0;
}
_$jscoverage['oauth.js'][42]++;
(function () {
  _$jscoverage['oauth.js'][44]++;
  var utils;
  _$jscoverage['oauth.js'][45]++;
  if (typeof require !== "undefined") {
    _$jscoverage['oauth.js'][46]++;
    utils = require("./utils");
  }
  else {
    _$jscoverage['oauth.js'][48]++;
    utils = weibo.utils;
  }
  _$jscoverage['oauth.js'][51]++;
  var OAuth = {};
  _$jscoverage['oauth.js'][53]++;
  OAuth.setProperties = (function setProperties(into, from) {
  _$jscoverage['oauth.js'][54]++;
  if (into && from) {
    _$jscoverage['oauth.js'][55]++;
    for (var key in from) {
      _$jscoverage['oauth.js'][56]++;
      into[key] = from[key];
}
  }
  _$jscoverage['oauth.js'][59]++;
  return into;
});
  _$jscoverage['oauth.js'][63]++;
  OAuth.setProperties(OAuth, {percentEncode: (function percentEncode(s) {
  _$jscoverage['oauth.js'][65]++;
  if (! s) {
    _$jscoverage['oauth.js'][66]++;
    return "";
  }
  _$jscoverage['oauth.js'][68]++;
  if (s instanceof Array) {
    _$jscoverage['oauth.js'][69]++;
    var e = "";
    _$jscoverage['oauth.js'][70]++;
    for (var i = 0; i < s.length; ++s) {
      _$jscoverage['oauth.js'][71]++;
      if (e) {
        _$jscoverage['oauth.js'][72]++;
        e += "&";
      }
      _$jscoverage['oauth.js'][74]++;
      e += percentEncode(s[i]);
}
    _$jscoverage['oauth.js'][76]++;
    return e;
  }
  _$jscoverage['oauth.js'][78]++;
  s = encodeURIComponent(s);
  _$jscoverage['oauth.js'][83]++;
  s = s.replace(/\!/g, "%21");
  _$jscoverage['oauth.js'][84]++;
  s = s.replace(/\*/g, "%2A");
  _$jscoverage['oauth.js'][85]++;
  s = s.replace(/\'/g, "%27");
  _$jscoverage['oauth.js'][87]++;
  s = s.replace(/\(/g, "%28");
  _$jscoverage['oauth.js'][88]++;
  s = s.replace(/\)/g, "%29");
  _$jscoverage['oauth.js'][89]++;
  return s;
}), decodePercent: decodeURIComponent, getParameterList: (function getParameterList(parameters) {
  _$jscoverage['oauth.js'][94]++;
  if (! parameters) {
    _$jscoverage['oauth.js'][95]++;
    return [];
  }
  _$jscoverage['oauth.js'][97]++;
  if (typeof parameters !== "object") {
    _$jscoverage['oauth.js'][98]++;
    return this.decodeForm(parameters + "");
  }
  _$jscoverage['oauth.js'][100]++;
  if (parameters instanceof Array) {
    _$jscoverage['oauth.js'][101]++;
    return parameters;
  }
  _$jscoverage['oauth.js'][103]++;
  var list = [];
  _$jscoverage['oauth.js'][104]++;
  for (var p in parameters) {
    _$jscoverage['oauth.js'][105]++;
    list.push([p, parameters[p]]);
}
  _$jscoverage['oauth.js'][107]++;
  return list;
}), getParameterMap: (function getParameterMap(parameters) {
  _$jscoverage['oauth.js'][111]++;
  if (! parameters) {
    _$jscoverage['oauth.js'][112]++;
    return ({});
  }
  _$jscoverage['oauth.js'][114]++;
  if (typeof parameters !== "object") {
    _$jscoverage['oauth.js'][115]++;
    return this.getParameterMap(this.decodeForm(parameters + ""));
  }
  _$jscoverage['oauth.js'][117]++;
  if (parameters instanceof Array) {
    _$jscoverage['oauth.js'][118]++;
    var map = {};
    _$jscoverage['oauth.js'][119]++;
    for (var p = 0; p < parameters.length; ++p) {
      _$jscoverage['oauth.js'][120]++;
      var key = parameters[p][0];
      _$jscoverage['oauth.js'][121]++;
      if (map[key] === undefined) {
        _$jscoverage['oauth.js'][122]++;
        map[key] = parameters[p][1];
      }
}
    _$jscoverage['oauth.js'][125]++;
    return map;
  }
  _$jscoverage['oauth.js'][127]++;
  return parameters;
}), formEncode: (function formEncode(parameters) {
  _$jscoverage['oauth.js'][130]++;
  var form = "";
  _$jscoverage['oauth.js'][131]++;
  var list = OAuth.getParameterList(parameters);
  _$jscoverage['oauth.js'][132]++;
  for (var p = 0, l = list.length; p < l; p++) {
    _$jscoverage['oauth.js'][133]++;
    var pair = list[p];
    _$jscoverage['oauth.js'][134]++;
    var value = pair[1];
    _$jscoverage['oauth.js'][135]++;
    if (! value) {
      _$jscoverage['oauth.js'][136]++;
      value = "";
    }
    _$jscoverage['oauth.js'][138]++;
    if (form) {
      _$jscoverage['oauth.js'][139]++;
      form += "&";
    }
    _$jscoverage['oauth.js'][141]++;
    form += OAuth.percentEncode(pair[0]) + "=" + OAuth.percentEncode(value);
}
  _$jscoverage['oauth.js'][143]++;
  return form;
}), decodeForm: (function decodeForm(form) {
  _$jscoverage['oauth.js'][146]++;
  var list = [];
  _$jscoverage['oauth.js'][147]++;
  var nvps = form.split("&");
  _$jscoverage['oauth.js'][148]++;
  for (var n = 0; n < nvps.length; ++n) {
    _$jscoverage['oauth.js'][149]++;
    var nvp = nvps[n];
    _$jscoverage['oauth.js'][150]++;
    if (! nvp) {
      _$jscoverage['oauth.js'][151]++;
      continue;
    }
    _$jscoverage['oauth.js'][153]++;
    var equals = nvp.indexOf("=");
    _$jscoverage['oauth.js'][154]++;
    var name;
    _$jscoverage['oauth.js'][155]++;
    var value;
    _$jscoverage['oauth.js'][156]++;
    if (equals < 0) {
      _$jscoverage['oauth.js'][157]++;
      name = OAuth.decodePercent(nvp);
      _$jscoverage['oauth.js'][158]++;
      value = null;
    }
    else {
      _$jscoverage['oauth.js'][160]++;
      name = OAuth.decodePercent(nvp.substring(0, equals));
      _$jscoverage['oauth.js'][161]++;
      value = OAuth.decodePercent(nvp.substring(equals + 1));
    }
    _$jscoverage['oauth.js'][163]++;
    list.push([name, value]);
}
  _$jscoverage['oauth.js'][165]++;
  return list;
}), setParameter: (function setParameter(message, name, value) {
  _$jscoverage['oauth.js'][168]++;
  var parameters = message.parameters;
  _$jscoverage['oauth.js'][169]++;
  if (parameters instanceof Array) {
    _$jscoverage['oauth.js'][170]++;
    for (var p = 0; p < parameters.length; ++p) {
      _$jscoverage['oauth.js'][171]++;
      if (parameters[p][0] === name) {
        _$jscoverage['oauth.js'][172]++;
        if (value === undefined) {
          _$jscoverage['oauth.js'][173]++;
          parameters.splice(p, 1);
        }
        else {
          _$jscoverage['oauth.js'][175]++;
          parameters[p][1] = value;
          _$jscoverage['oauth.js'][176]++;
          value = undefined;
        }
      }
}
    _$jscoverage['oauth.js'][180]++;
    if (value !== undefined) {
      _$jscoverage['oauth.js'][181]++;
      parameters.push([name, value]);
    }
  }
  else {
    _$jscoverage['oauth.js'][184]++;
    parameters = OAuth.getParameterMap(parameters);
    _$jscoverage['oauth.js'][185]++;
    parameters[name] = value;
    _$jscoverage['oauth.js'][186]++;
    message.parameters = parameters;
  }
}), setParameters: (function setParameters(message, parameters) {
  _$jscoverage['oauth.js'][190]++;
  var list = OAuth.getParameterList(parameters);
  _$jscoverage['oauth.js'][191]++;
  for (var i = 0; i < list.length; ++i) {
    _$jscoverage['oauth.js'][192]++;
    OAuth.setParameter(message, list[i][0], list[i][1]);
}
}), setTimestampAndNonce: (function setTimestampAndNonce(message) {
  _$jscoverage['oauth.js'][196]++;
  OAuth.setParameter(message, "oauth_timestamp", OAuth.timestamp());
  _$jscoverage['oauth.js'][197]++;
  OAuth.setParameter(message, "oauth_nonce", OAuth.nonce(32));
}), addToURL: (function addToURL(url, parameters) {
  _$jscoverage['oauth.js'][200]++;
  if (parameters) {
    _$jscoverage['oauth.js'][201]++;
    var toAdd = OAuth.formEncode(parameters);
    _$jscoverage['oauth.js'][202]++;
    if (toAdd) {
      _$jscoverage['oauth.js'][203]++;
      if (url.indexOf("?") < 0) {
        _$jscoverage['oauth.js'][204]++;
        url += "?";
      }
      else {
        _$jscoverage['oauth.js'][206]++;
        url += "&";
      }
      _$jscoverage['oauth.js'][208]++;
      url += toAdd;
    }
  }
  _$jscoverage['oauth.js'][211]++;
  return url;
}), getAuthorizationHeader: (function getAuthorizationHeader(realm, parameters) {
  _$jscoverage['oauth.js'][215]++;
  var header = "";
  _$jscoverage['oauth.js'][216]++;
  if (realm) {
    _$jscoverage['oauth.js'][217]++;
    header += ", realm=\"" + OAuth.percentEncode(realm) + "\"";
  }
  _$jscoverage['oauth.js'][219]++;
  var list = OAuth.getParameterList(parameters);
  _$jscoverage['oauth.js'][220]++;
  for (var p = 0; p < list.length; ++p) {
    _$jscoverage['oauth.js'][221]++;
    var parameter = list[p];
    _$jscoverage['oauth.js'][222]++;
    var name = parameter[0];
    _$jscoverage['oauth.js'][223]++;
    if (name.indexOf("oauth_") === 0) {
      _$jscoverage['oauth.js'][224]++;
      header += ", " + OAuth.percentEncode(name) + "=\"" + OAuth.percentEncode(parameter[1]) + "\"";
    }
}
  _$jscoverage['oauth.js'][227]++;
  return "OAuth " + header.substring(2);
}), timestamp: (function timestamp() {
  _$jscoverage['oauth.js'][231]++;
  return Math.floor(new Date().getTime() / 1000);
}), nonce: (function nonce(length) {
  _$jscoverage['oauth.js'][235]++;
  if (! length) {
    _$jscoverage['oauth.js'][236]++;
    return "";
  }
  _$jscoverage['oauth.js'][238]++;
  var chars = OAuth.nonce.CHARS;
  _$jscoverage['oauth.js'][239]++;
  var result = "";
  _$jscoverage['oauth.js'][240]++;
  for (var i = 0; i < length; ++i) {
    _$jscoverage['oauth.js'][241]++;
    var rnum = Math.floor(Math.random() * chars.length);
    _$jscoverage['oauth.js'][242]++;
    result += chars.substring(rnum, rnum + 1);
}
  _$jscoverage['oauth.js'][244]++;
  return result;
})});
  _$jscoverage['oauth.js'][248]++;
  OAuth.nonce.CHARS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXTZabcdefghiklmnopqrstuvwxyz";
  _$jscoverage['oauth.js'][254]++;
  OAuth.declareClass = (function declareClass(parent, name, newConstructor) {
  _$jscoverage['oauth.js'][255]++;
  var previous = parent[name];
  _$jscoverage['oauth.js'][256]++;
  parent[name] = newConstructor;
  _$jscoverage['oauth.js'][257]++;
  if (newConstructor && previous) {
    _$jscoverage['oauth.js'][258]++;
    for (var key in previous) {
      _$jscoverage['oauth.js'][259]++;
      if (key !== "prototype") {
        _$jscoverage['oauth.js'][260]++;
        newConstructor[key] = previous[key];
      }
}
  }
  _$jscoverage['oauth.js'][264]++;
  return newConstructor;
});
  _$jscoverage['oauth.js'][268]++;
  OAuth.declareClass(OAuth, "SignatureMethod", (function OAuthSignatureMethod() {
}));
  _$jscoverage['oauth.js'][271]++;
  OAuth.setProperties(OAuth.SignatureMethod.prototype, {sign: (function sign(message) {
  _$jscoverage['oauth.js'][274]++;
  var baseString = OAuth.SignatureMethod.getBaseString(message);
  _$jscoverage['oauth.js'][275]++;
  var signature = this.getSignature(baseString);
  _$jscoverage['oauth.js'][277]++;
  OAuth.setParameter(message, "oauth_signature", signature);
  _$jscoverage['oauth.js'][278]++;
  return signature;
}), initialize: (function initialize(name, accessor) {
  _$jscoverage['oauth.js'][282]++;
  var consumerSecret;
  _$jscoverage['oauth.js'][283]++;
  if (accessor.accessorSecret && name.length > 9 && name.substring(name.length - 9) === "-Accessor") {
    _$jscoverage['oauth.js'][284]++;
    consumerSecret = accessor.accessorSecret;
  }
  else {
    _$jscoverage['oauth.js'][286]++;
    consumerSecret = accessor.consumerSecret;
  }
  _$jscoverage['oauth.js'][288]++;
  this.key = OAuth.percentEncode(consumerSecret) + "&" + OAuth.percentEncode(accessor.tokenSecret);
})});
  _$jscoverage['oauth.js'][297]++;
  OAuth.setProperties(OAuth.SignatureMethod, {sign: (function sign(message, accessor) {
  _$jscoverage['oauth.js'][299]++;
  var name = OAuth.getParameterMap(message.parameters).oauth_signature_method;
  _$jscoverage['oauth.js'][300]++;
  if (! name) {
    _$jscoverage['oauth.js'][301]++;
    name = "HMAC-SHA1";
    _$jscoverage['oauth.js'][302]++;
    OAuth.setParameter(message, "oauth_signature_method", name);
  }
  _$jscoverage['oauth.js'][304]++;
  OAuth.SignatureMethod.newMethod(name, accessor).sign(message);
}), newMethod: (function newMethod(name, accessor) {
  _$jscoverage['oauth.js'][309]++;
  var Impl = OAuth.SignatureMethod.REGISTERED[name];
  _$jscoverage['oauth.js'][310]++;
  if (typeof Impl === "function") {
    _$jscoverage['oauth.js'][311]++;
    var method = new Impl();
    _$jscoverage['oauth.js'][312]++;
    method.initialize(name, accessor);
    _$jscoverage['oauth.js'][313]++;
    return method;
  }
  _$jscoverage['oauth.js'][315]++;
  var err = new Error("signature_method_rejected");
  _$jscoverage['oauth.js'][316]++;
  var acceptable = "";
  _$jscoverage['oauth.js'][317]++;
  for (var r in OAuth.SignatureMethod.REGISTERED) {
    _$jscoverage['oauth.js'][318]++;
    if (acceptable) {
      _$jscoverage['oauth.js'][319]++;
      acceptable += "&";
    }
    _$jscoverage['oauth.js'][321]++;
    acceptable += OAuth.percentEncode(r);
}
  _$jscoverage['oauth.js'][323]++;
  err.oauth_acceptable_signature_methods = acceptable;
  _$jscoverage['oauth.js'][324]++;
  throw err;
}), REGISTERED: {}, registerMethodClass: (function registerMethodClass(names, classConstructor) {
  _$jscoverage['oauth.js'][334]++;
  for (var n = 0, l = names.length; n < l; ++n) {
    _$jscoverage['oauth.js'][335]++;
    OAuth.SignatureMethod.REGISTERED[names[n]] = classConstructor;
}
}), makeSubclass: (function makeSubclass(getSignatureFunction) {
  _$jscoverage['oauth.js'][340]++;
  var SuperClass = OAuth.SignatureMethod;
  _$jscoverage['oauth.js'][341]++;
  var subClass = (function () {
  _$jscoverage['oauth.js'][342]++;
  SuperClass.call(this);
});
  _$jscoverage['oauth.js'][344]++;
  subClass.prototype = new SuperClass();
  _$jscoverage['oauth.js'][347]++;
  subClass.prototype.getSignature = getSignatureFunction;
  _$jscoverage['oauth.js'][348]++;
  subClass.prototype.constructor = subClass;
  _$jscoverage['oauth.js'][349]++;
  return subClass;
}), getBaseString: (function getBaseString(message) {
  _$jscoverage['oauth.js'][352]++;
  var URL = message.action;
  _$jscoverage['oauth.js'][353]++;
  var q = URL.indexOf("?");
  _$jscoverage['oauth.js'][354]++;
  var parameters;
  _$jscoverage['oauth.js'][355]++;
  if (q < 0) {
    _$jscoverage['oauth.js'][356]++;
    parameters = message.parameters;
  }
  else {
    _$jscoverage['oauth.js'][359]++;
    parameters = OAuth.decodeForm(URL.substring(q + 1));
    _$jscoverage['oauth.js'][360]++;
    var toAdd = OAuth.getParameterList(message.parameters);
    _$jscoverage['oauth.js'][361]++;
    for (var a = 0, l = toAdd.length; a < l; ++a) {
      _$jscoverage['oauth.js'][362]++;
      parameters.push(toAdd[a]);
}
  }
  _$jscoverage['oauth.js'][365]++;
  return OAuth.percentEncode(message.method.toUpperCase()) + "&" + OAuth.percentEncode(OAuth.SignatureMethod.normalizeUrl(URL)) + "&" + OAuth.percentEncode(OAuth.SignatureMethod.normalizeParameters(parameters));
}), normalizeUrl: (function normalizeUrl(url) {
  _$jscoverage['oauth.js'][370]++;
  var uri = OAuth.SignatureMethod.parseUri(url);
  _$jscoverage['oauth.js'][371]++;
  var scheme = uri.protocol.toLowerCase();
  _$jscoverage['oauth.js'][372]++;
  var authority = uri.authority.toLowerCase();
  _$jscoverage['oauth.js'][373]++;
  var dropPort = (scheme === "http" && uri.port === 80) || (scheme === "https" && uri.port === 443);
  _$jscoverage['oauth.js'][374]++;
  if (dropPort) {
    _$jscoverage['oauth.js'][376]++;
    var index = authority.lastIndexOf(":");
    _$jscoverage['oauth.js'][377]++;
    if (index >= 0) {
      _$jscoverage['oauth.js'][378]++;
      authority = authority.substring(0, index);
    }
  }
  _$jscoverage['oauth.js'][381]++;
  var path = uri.path;
  _$jscoverage['oauth.js'][386]++;
  return scheme + "://" + authority + path;
}), parseUri: (function parseUri(str) {
  _$jscoverage['oauth.js'][392]++;
  var o = {key: ["source", "protocol", "authority", "userInfo", "user", "password", "host", "port", "relative", "path", "directory", "file", "query", "anchor"], parser: {strict: /^(?:([^:\/?#]+):)?(?:\/\/((?:(([^:@]*):?([^:@]*))?@)?([^:\/?#]*)(?::(\d*))?))?((((?:[^?#\/]*\/)*)([^?#]*))(?:\?([^#]*))?(?:#(.*))?)/}};
  _$jscoverage['oauth.js'][395]++;
  var m = o.parser.strict.exec(str);
  _$jscoverage['oauth.js'][396]++;
  var uri = {};
  _$jscoverage['oauth.js'][397]++;
  var i = 14;
  _$jscoverage['oauth.js'][398]++;
  while (i--) {
    _$jscoverage['oauth.js'][399]++;
    uri[o.key[i]] = m[i] || "";
}
  _$jscoverage['oauth.js'][401]++;
  return uri;
}), normalizeParameters: (function normalizeParameters(parameters) {
  _$jscoverage['oauth.js'][404]++;
  if (! parameters) {
    _$jscoverage['oauth.js'][405]++;
    return "";
  }
  _$jscoverage['oauth.js'][407]++;
  var norm = [];
  _$jscoverage['oauth.js'][408]++;
  var list = OAuth.getParameterList(parameters);
  _$jscoverage['oauth.js'][409]++;
  for (var p = 0; p < list.length; ++p) {
    _$jscoverage['oauth.js'][410]++;
    var nvp = list[p];
    _$jscoverage['oauth.js'][411]++;
    if (nvp[0] !== "oauth_signature") {
      _$jscoverage['oauth.js'][412]++;
      norm.push(nvp);
    }
}
  _$jscoverage['oauth.js'][415]++;
  norm.sort((function (a, b) {
  _$jscoverage['oauth.js'][416]++;
  if (a[0] < b[0]) {
    _$jscoverage['oauth.js'][416]++;
    return -1;
  }
  _$jscoverage['oauth.js'][417]++;
  if (a[0] > b[0]) {
    _$jscoverage['oauth.js'][417]++;
    return 1;
  }
  _$jscoverage['oauth.js'][418]++;
  if (a[1] < b[1]) {
    _$jscoverage['oauth.js'][418]++;
    return -1;
  }
  _$jscoverage['oauth.js'][419]++;
  if (a[1] > b[1]) {
    _$jscoverage['oauth.js'][419]++;
    return 1;
  }
  _$jscoverage['oauth.js'][420]++;
  return 0;
}));
  _$jscoverage['oauth.js'][422]++;
  return OAuth.formEncode(norm);
})});
  _$jscoverage['oauth.js'][426]++;
  OAuth.SignatureMethod.registerMethodClass(["PLAINTEXT", "PLAINTEXT-Accessor"], OAuth.SignatureMethod.makeSubclass((function getSignature(baseString) {
  _$jscoverage['oauth.js'][429]++;
  return this.key;
})));
  _$jscoverage['oauth.js'][433]++;
  OAuth.SignatureMethod.registerMethodClass(["HMAC-SHA1", "HMAC-SHA1-Accessor"], OAuth.SignatureMethod.makeSubclass((function getSignature(baseString) {
  _$jscoverage['oauth.js'][436]++;
  return utils.base64HmacSha1(baseString, this.key);
})));
  _$jscoverage['oauth.js'][440]++;
  var root = this;
  _$jscoverage['oauth.js'][441]++;
  if (typeof module === "undefined") {
    _$jscoverage['oauth.js'][442]++;
    root.weibo = root.weibo || {};
    _$jscoverage['oauth.js'][443]++;
    root.weibo.OAuth = OAuth;
  }
  else {
    _$jscoverage['oauth.js'][445]++;
    module.exports = OAuth;
  }
})();
_$jscoverage['oauth.js'].source = ["/*"," * Copyright 2008 Netflix, Inc."," *"," * Licensed under the Apache License, Version 2.0 (the \"License\");"," * you may not use this file except in compliance with the License."," * You may obtain a copy of the License at"," *"," *     http://www.apache.org/licenses/LICENSE-2.0"," *"," * Unless required by applicable law or agreed to in writing, software"," * distributed under the License is distributed on an \"AS IS\" BASIS,"," * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied."," * See the License for the specific language governing permissions and"," * limitations under the License."," */","// Here's some JavaScript software that's useful for implementing OAuth.","// The HMAC-SHA1 signature method calls b64_hmac_sha1, defined by","// http://pajhome.org.uk/crypt/md5/sha1.js","/* An OAuth message is represented as an object like this:","   {method: \"GET\", action: \"http://server.com/path\", parameters: ...}","   The parameters may be either a map {name: value, name2: value2}","   or an Array of name-value pairs [[name, value], [name2, value2]].","   The latter representation is more powerful: it supports parameters","   in a specific sequence, or several parameters with the same name;","   for example [[\"a\", 1], [\"b\", 2], [\"a\", 3]].","   Parameter names and values are NOT percent-encoded in an object.","   They must be encoded before transmission and decoded after reception.","   For example, this message object:","   {method: \"GET\", action: \"http://server/path\", parameters: {p: \"x y\"}}","   ... can be transmitted as an HTTP request that begins:","   GET /path?p=x%20y HTTP/1.0","   (This isn't a valid OAuth request, since it lacks a signature etc.)","   Note that the object \"x y\" is transmitted as x%20y.  To encode","   parameters, you can call OAuth.addToURL, OAuth.formEncode or","   OAuth.getAuthorization.","   This message object model harmonizes with the browser object model for","   input elements of an form, whose value property isn't percent encoded.","   The browser encodes each value before transmitting it. For example,","   see consumer.setInputs in example/consumer.js."," */","","(function () {","","var utils;","if (typeof require !== 'undefined') {","  utils = require('./utils');","} else {","  utils = weibo.utils;","}","","var OAuth = {};","","OAuth.setProperties = function setProperties(into, from) {","  if (into &amp;&amp; from) {","    for (var key in from) {","      into[key] = from[key];","    }","  }","  return into;","};","","// utility functions","OAuth.setProperties(OAuth, {","  percentEncode: function percentEncode(s) {","    if (!s) {","      return \"\";","    }","    if (s instanceof Array) {","      var e = \"\";","      for (var i = 0; i &lt; s.length; ++s) {","        if (e) {","          e += '&amp;';","        }","        e += percentEncode(s[i]);","      }","      return e;","    }","    s = encodeURIComponent(s);","    // Now replace the values which encodeURIComponent doesn't do","    // encodeURIComponent ignores: - _ . ! ~ * ' ( )","    // OAuth dictates the only ones you can ignore are: - _ . ~","    // Source: http://developer.mozilla.org/en/docs/Core_JavaScript_1.5_Reference:Global_Functions:encodeURIComponent","    s = s.replace(/\\!/g, \"%21\");","    s = s.replace(/\\*/g, \"%2A\");","    s = s.replace(/\\'/g, \"%27\");","    // s = s.replace(\"(\", \"%28\", \"g\");","    s = s.replace(/\\(/g, \"%28\");","    s = s.replace(/\\)/g, \"%29\");","    return s;","  },","  decodePercent: decodeURIComponent,","  /** Convert the given parameters to an Array of name-value pairs. */","  getParameterList: function getParameterList(parameters) {","    if (!parameters) {","      return [];","    }","    if (typeof parameters !== \"object\") {","      return this.decodeForm(parameters + \"\");","    }","    if (parameters instanceof Array) {","      return parameters;","    }","    var list = [];","    for (var p in parameters) {","      list.push([ p, parameters[p] ]);","    }","    return list;","  },","  /** Convert the given parameters to a map from name to value. */","  getParameterMap: function getParameterMap(parameters) {","    if (!parameters) {","      return {};","    }","    if (typeof parameters !== \"object\") {","      return this.getParameterMap(this.decodeForm(parameters + \"\"));","    }","    if (parameters instanceof Array) {","      var map = {};","      for (var p = 0; p &lt; parameters.length; ++p) {","        var key = parameters[p][0];","        if (map[key] === undefined) { // first value wins","          map[key] = parameters[p][1];","        }","      }","      return map;","    }","    return parameters;","  },","  formEncode: function formEncode(parameters) {","    var form = \"\";","    var list = OAuth.getParameterList(parameters);","    for (var p = 0, l = list.length; p &lt; l; p++) {","      var pair = list[p];","      var value = pair[1];","      if (!value) {","        value = \"\";","      }","      if (form) {","        form += '&amp;';","      }","      form += OAuth.percentEncode(pair[0]) + '=' + OAuth.percentEncode(value);","    }","    return form;","  },","  decodeForm: function decodeForm(form) {","    var list = [];","    var nvps = form.split('&amp;');","    for (var n = 0; n &lt; nvps.length; ++n) {","      var nvp = nvps[n];","      if (!nvp) {","        continue;","      }","      var equals = nvp.indexOf('=');","      var name;","      var value;","      if (equals &lt; 0) {","        name = OAuth.decodePercent(nvp);","        value = null;","      } else {","        name = OAuth.decodePercent(nvp.substring(0, equals));","        value = OAuth.decodePercent(nvp.substring(equals + 1));","      }","      list.push([name, value]);","    }","    return list;","  },","  setParameter: function setParameter(message, name, value) {","    var parameters = message.parameters;","    if (parameters instanceof Array) {","      for (var p = 0; p &lt; parameters.length; ++p) {","        if (parameters[p][0] === name) {","          if (value === undefined) {","            parameters.splice(p, 1);","          } else {","            parameters[p][1] = value;","            value = undefined;","          }","        }","      }","      if (value !== undefined) {","        parameters.push([name, value]);","      }","    } else {","      parameters = OAuth.getParameterMap(parameters);","      parameters[name] = value;","      message.parameters = parameters;","    }","  },","  setParameters: function setParameters(message, parameters) {","    var list = OAuth.getParameterList(parameters);","    for (var i = 0; i &lt; list.length; ++i) {","      OAuth.setParameter(message, list[i][0], list[i][1]);","    }","  },","  setTimestampAndNonce: function setTimestampAndNonce(message) {","    OAuth.setParameter(message, \"oauth_timestamp\", OAuth.timestamp());","    OAuth.setParameter(message, \"oauth_nonce\", OAuth.nonce(32));","  },","  addToURL: function addToURL(url, parameters) {","      if (parameters) {","        var toAdd = OAuth.formEncode(parameters);","        if (toAdd) {","          if (url.indexOf('?') &lt; 0) {","            url += '?';","          } else {","            url += '&amp;';","          }","          url += toAdd;","        }","      }","      return url;","  },","  /** Construct the value of the Authorization header for an HTTP request. */","  getAuthorizationHeader: function getAuthorizationHeader(realm, parameters) {","    var header = '';","    if (realm) {","      header += ', realm=\"' + OAuth.percentEncode(realm) + '\"';","    }","    var list = OAuth.getParameterList(parameters);","    for (var p = 0; p &lt; list.length; ++p) {","      var parameter = list[p];","      var name = parameter[0];","      if (name.indexOf(\"oauth_\") === 0) {","        header += ', ' + OAuth.percentEncode(name) + '=\"' + OAuth.percentEncode(parameter[1]) + '\"';","      }","    }","    return 'OAuth ' + header.substring(2);","  },","","  timestamp: function timestamp() {","    return Math.floor(new Date().getTime() / 1000);","  },","","  nonce: function nonce(length) {","    if (!length) {","      return '';","    }","    var chars = OAuth.nonce.CHARS;","    var result = \"\";","    for (var i = 0; i &lt; length; ++i) {","      var rnum = Math.floor(Math.random() * chars.length);","      result += chars.substring(rnum, rnum + 1);","    }","    return result;","  }","});","","OAuth.nonce.CHARS = \"0123456789ABCDEFGHIJKLMNOPQRSTUVWXTZabcdefghiklmnopqrstuvwxyz\";","/** Define a constructor function,","    without causing trouble to anyone who was using it as a namespace.","    That is, if parent[name] already existed and had properties,","    copy those properties into the new constructor."," */","OAuth.declareClass = function declareClass(parent, name, newConstructor) {","  var previous = parent[name];","  parent[name] = newConstructor;","  if (newConstructor &amp;&amp; previous) {","    for (var key in previous) {","      if (key !== \"prototype\") {","        newConstructor[key] = previous[key];","      }","    }","  }","  return newConstructor;","};","","/** An abstract algorithm for signing messages. */","OAuth.declareClass(OAuth, \"SignatureMethod\", function OAuthSignatureMethod() {});","","// instance members","OAuth.setProperties(OAuth.SignatureMethod.prototype, {","  /** Add a signature to the message. */","  sign: function sign(message) {","    var baseString = OAuth.SignatureMethod.getBaseString(message);","    var signature = this.getSignature(baseString);","    // console.log(baseString, this.key, signature)","    OAuth.setParameter(message, \"oauth_signature\", signature);","    return signature; // just in case someone's interested","  },","  /** Set the key string for signing. */","  initialize: function initialize(name, accessor) {","    var consumerSecret;","    if (accessor.accessorSecret &amp;&amp; name.length &gt; 9 &amp;&amp; name.substring(name.length-9) === \"-Accessor\") {","        consumerSecret = accessor.accessorSecret;","    } else {","        consumerSecret = accessor.consumerSecret;","    }","    this.key = OAuth.percentEncode(consumerSecret) + \"&amp;\" + OAuth.percentEncode(accessor.tokenSecret);","  }","});","","/* SignatureMethod expects an accessor object to be like this:","   {tokenSecret: \"lakjsdflkj...\", consumerSecret: \"QOUEWRI..\", accessorSecret: \"xcmvzc...\"}","   The accessorSecret property is optional."," */","// Class members:","OAuth.setProperties(OAuth.SignatureMethod, {","  sign: function sign(message, accessor) {","    var name = OAuth.getParameterMap(message.parameters).oauth_signature_method;","    if (!name) {","      name = 'HMAC-SHA1';","      OAuth.setParameter(message, 'oauth_signature_method', name);","    }","    OAuth.SignatureMethod.newMethod(name, accessor).sign(message);","  },","","  /** Instantiate a SignatureMethod for the given method name. */","  newMethod: function newMethod(name, accessor) {","    var Impl = OAuth.SignatureMethod.REGISTERED[name];","    if (typeof Impl === 'function') {","      var method = new Impl();","      method.initialize(name, accessor);","      return method;","    }","    var err = new Error(\"signature_method_rejected\");","    var acceptable = \"\";","    for (var r in OAuth.SignatureMethod.REGISTERED) {","      if (acceptable) {","        acceptable += '&amp;';","      }","      acceptable += OAuth.percentEncode(r);","    }","    err.oauth_acceptable_signature_methods = acceptable;","    throw err;","  },","  /** A map from signature method name to constructor. */","  REGISTERED: {},","  /** Subsequently, the given constructor will be used for the named methods.","      The constructor will be called with no parameters.","      The resulting object should usually implement getSignature(baseString).","      You can easily define such a constructor by calling makeSubclass, below.","   */","  registerMethodClass: function registerMethodClass(names, classConstructor) {","    for (var n = 0, l = names.length; n &lt; l; ++n) {","      OAuth.SignatureMethod.REGISTERED[names[n]] = classConstructor;","    }","  },","  /** Create a subclass of OAuth.SignatureMethod, with the given getSignature function. */","  makeSubclass: function makeSubclass(getSignatureFunction) {","    var SuperClass = OAuth.SignatureMethod;","    var subClass = function() {","      SuperClass.call(this);","    }; ","    subClass.prototype = new SuperClass();","    // Delete instance variables from prototype:","    // delete subclass.prototype... There aren't any.","    subClass.prototype.getSignature = getSignatureFunction;","    subClass.prototype.constructor = subClass;","    return subClass;","  },","  getBaseString: function getBaseString(message) {","      var URL = message.action;","      var q = URL.indexOf('?');","      var parameters;","      if (q &lt; 0) {","          parameters = message.parameters;","      } else {","          // Combine the URL query string with the other parameters:","          parameters = OAuth.decodeForm(URL.substring(q + 1));","          var toAdd = OAuth.getParameterList(message.parameters);","          for (var a = 0, l = toAdd.length; a &lt; l; ++a) {","              parameters.push(toAdd[a]);","          }","      }","      return OAuth.percentEncode(message.method.toUpperCase()) + '&amp;' +","        OAuth.percentEncode(OAuth.SignatureMethod.normalizeUrl(URL)) + '&amp;' +","        OAuth.percentEncode(OAuth.SignatureMethod.normalizeParameters(parameters));","  },","  normalizeUrl: function normalizeUrl(url) {","    var uri = OAuth.SignatureMethod.parseUri(url);","    var scheme = uri.protocol.toLowerCase();","    var authority = uri.authority.toLowerCase();","    var dropPort = (scheme === \"http\" &amp;&amp; uri.port === 80) || (scheme === \"https\" &amp;&amp; uri.port === 443);","    if (dropPort) {","      // find the last : in the authority","      var index = authority.lastIndexOf(\":\");","      if (index &gt;= 0) {","        authority = authority.substring(0, index);","      }","    }","    var path = uri.path;","//        if (!path) {","//            path = \"/\"; // conforms to RFC 2616 section 3.2.2","//        }","    // we know that there is no query and no fragment here.","    return scheme + \"://\" + authority + path;","  },","  parseUri: function parseUri(str) {","    /* This function was adapted from parseUri 1.2.1","       http://stevenlevithan.com/demo/parseuri/js/assets/parseuri.js","     */","    var o = {","      key: [ \"source\",\"protocol\",\"authority\",\"userInfo\",\"user\",\"password\",\"host\",\"port\",\"relative\",\"path\",\"directory\",\"file\",\"query\",\"anchor\"],","      parser: { strict: /^(?:([^:\\/?#]+):)?(?:\\/\\/((?:(([^:@]*):?([^:@]*))?@)?([^:\\/?#]*)(?::(\\d*))?))?((((?:[^?#\\/]*\\/)*)([^?#]*))(?:\\?([^#]*))?(?:#(.*))?)/ }};","    var m = o.parser.strict.exec(str);","    var uri = {};","    var i = 14;","    while (i--) {","      uri[o.key[i]] = m[i] || \"\";","    }","    return uri;","  },","  normalizeParameters: function normalizeParameters(parameters) {","    if (!parameters) {","      return \"\";","    }","    var norm = [];","    var list = OAuth.getParameterList(parameters);","    for (var p = 0; p &lt; list.length; ++p) {","      var nvp = list[p];","      if (nvp[0] !== \"oauth_signature\") {","        norm.push(nvp);","      }","    }","    norm.sort(function (a, b) {","      if (a[0] &lt; b[0]) { return -1; }","      if (a[0] &gt; b[0]) { return 1; }","      if (a[1] &lt; b[1]) { return  -1; }","      if (a[1] &gt; b[1]) { return 1; }","      return 0;","    });","    return OAuth.formEncode(norm);","  }","});","","OAuth.SignatureMethod.registerMethodClass([\"PLAINTEXT\", \"PLAINTEXT-Accessor\"],","OAuth.SignatureMethod.makeSubclass(","  function getSignature(baseString) {","    return this.key;","  }","));","","OAuth.SignatureMethod.registerMethodClass([\"HMAC-SHA1\", \"HMAC-SHA1-Accessor\"],","OAuth.SignatureMethod.makeSubclass(","  function getSignature(baseString) {","    return utils.base64HmacSha1(baseString, this.key);","  }","));","","var root = this; // window on browser","if (typeof module === 'undefined') {","  root.weibo = root.weibo || {};","  root.weibo.OAuth = OAuth;","} else {","  module.exports = OAuth;","}","","})();"];
