// weibo/lib-cov/tbase_oauth_v2.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['tbase_oauth_v2.js']) {
  _$jscoverage['tbase_oauth_v2.js'] = [];
  _$jscoverage['tbase_oauth_v2.js'][7] = 0;
  _$jscoverage['tbase_oauth_v2.js'][13] = 0;
  _$jscoverage['tbase_oauth_v2.js'][14] = 0;
  _$jscoverage['tbase_oauth_v2.js'][15] = 0;
  _$jscoverage['tbase_oauth_v2.js'][16] = 0;
  _$jscoverage['tbase_oauth_v2.js'][21] = 0;
  _$jscoverage['tbase_oauth_v2.js'][22] = 0;
  _$jscoverage['tbase_oauth_v2.js'][23] = 0;
  _$jscoverage['tbase_oauth_v2.js'][26] = 0;
  _$jscoverage['tbase_oauth_v2.js'][27] = 0;
  _$jscoverage['tbase_oauth_v2.js'][33] = 0;
  _$jscoverage['tbase_oauth_v2.js'][34] = 0;
  _$jscoverage['tbase_oauth_v2.js'][35] = 0;
  _$jscoverage['tbase_oauth_v2.js'][42] = 0;
  _$jscoverage['tbase_oauth_v2.js'][43] = 0;
  _$jscoverage['tbase_oauth_v2.js'][48] = 0;
  _$jscoverage['tbase_oauth_v2.js'][49] = 0;
  _$jscoverage['tbase_oauth_v2.js'][50] = 0;
  _$jscoverage['tbase_oauth_v2.js'][52] = 0;
  _$jscoverage['tbase_oauth_v2.js'][54] = 0;
  _$jscoverage['tbase_oauth_v2.js'][56] = 0;
  _$jscoverage['tbase_oauth_v2.js'][59] = 0;
  _$jscoverage['tbase_oauth_v2.js'][60] = 0;
  _$jscoverage['tbase_oauth_v2.js'][61] = 0;
  _$jscoverage['tbase_oauth_v2.js'][62] = 0;
  _$jscoverage['tbase_oauth_v2.js'][66] = 0;
  _$jscoverage['tbase_oauth_v2.js'][67] = 0;
  _$jscoverage['tbase_oauth_v2.js'][69] = 0;
  _$jscoverage['tbase_oauth_v2.js'][72] = 0;
  _$jscoverage['tbase_oauth_v2.js'][73] = 0;
  _$jscoverage['tbase_oauth_v2.js'][80] = 0;
  _$jscoverage['tbase_oauth_v2.js'][81] = 0;
  _$jscoverage['tbase_oauth_v2.js'][82] = 0;
  _$jscoverage['tbase_oauth_v2.js'][83] = 0;
  _$jscoverage['tbase_oauth_v2.js'][84] = 0;
  _$jscoverage['tbase_oauth_v2.js'][85] = 0;
  _$jscoverage['tbase_oauth_v2.js'][88] = 0;
  _$jscoverage['tbase_oauth_v2.js'][89] = 0;
  _$jscoverage['tbase_oauth_v2.js'][90] = 0;
  _$jscoverage['tbase_oauth_v2.js'][91] = 0;
  _$jscoverage['tbase_oauth_v2.js'][92] = 0;
  _$jscoverage['tbase_oauth_v2.js'][93] = 0;
  _$jscoverage['tbase_oauth_v2.js'][99] = 0;
  _$jscoverage['tbase_oauth_v2.js'][100] = 0;
  _$jscoverage['tbase_oauth_v2.js'][101] = 0;
  _$jscoverage['tbase_oauth_v2.js'][102] = 0;
  _$jscoverage['tbase_oauth_v2.js'][103] = 0;
  _$jscoverage['tbase_oauth_v2.js'][104] = 0;
  _$jscoverage['tbase_oauth_v2.js'][105] = 0;
  _$jscoverage['tbase_oauth_v2.js'][107] = 0;
  _$jscoverage['tbase_oauth_v2.js'][108] = 0;
  _$jscoverage['tbase_oauth_v2.js'][110] = 0;
  _$jscoverage['tbase_oauth_v2.js'][113] = 0;
  _$jscoverage['tbase_oauth_v2.js'][114] = 0;
  _$jscoverage['tbase_oauth_v2.js'][115] = 0;
}
_$jscoverage['tbase_oauth_v2.js'][7]++;
"use strict";
_$jscoverage['tbase_oauth_v2.js'][13]++;
var TBase = require("./tbase");
_$jscoverage['tbase_oauth_v2.js'][14]++;
var inherits = require("util").inherits;
_$jscoverage['tbase_oauth_v2.js'][15]++;
var utils = require("./utils");
_$jscoverage['tbase_oauth_v2.js'][16]++;
var querystring = require("querystring");
_$jscoverage['tbase_oauth_v2.js'][21]++;
function TBaseOauthV2() {
  _$jscoverage['tbase_oauth_v2.js'][22]++;
  TBaseOauthV2.super_.call(this);
  _$jscoverage['tbase_oauth_v2.js'][23]++;
  this.config.oauth_version = "2.0";
}
_$jscoverage['tbase_oauth_v2.js'][26]++;
inherits(TBaseOauthV2, TBase);
_$jscoverage['tbase_oauth_v2.js'][27]++;
module.exports = TBaseOauthV2;
_$jscoverage['tbase_oauth_v2.js'][33]++;
TBaseOauthV2.prototype.format_access_token = (function (token) {
  _$jscoverage['tbase_oauth_v2.js'][34]++;
  token = JSON.parse(token);
  _$jscoverage['tbase_oauth_v2.js'][35]++;
  return token;
});
_$jscoverage['tbase_oauth_v2.js'][42]++;
TBaseOauthV2.prototype.convert_token = (function (user) {
  _$jscoverage['tbase_oauth_v2.js'][43]++;
  var params = {redirect_uri: user.oauth_callback || this.config.oauth_callback, client_id: this.config.appkey, response_type: "code"};
  _$jscoverage['tbase_oauth_v2.js'][48]++;
  var oauth_scope = user.oauth_scope || this.config.oauth_scope;
  _$jscoverage['tbase_oauth_v2.js'][49]++;
  if (oauth_scope) {
    _$jscoverage['tbase_oauth_v2.js'][50]++;
    params.oauth_scope = oauth_scope;
  }
  _$jscoverage['tbase_oauth_v2.js'][52]++;
  if (user.state) {
    _$jscoverage['tbase_oauth_v2.js'][54]++;
    params.state = user.state;
  }
  _$jscoverage['tbase_oauth_v2.js'][56]++;
  return params;
});
_$jscoverage['tbase_oauth_v2.js'][59]++;
TBaseOauthV2.prototype.get_authorization_url = (function (user, callback) {
  _$jscoverage['tbase_oauth_v2.js'][60]++;
  var data = this.convert_token(user);
  _$jscoverage['tbase_oauth_v2.js'][61]++;
  data.response_type = "code";
  _$jscoverage['tbase_oauth_v2.js'][62]++;
  var info = {blogtype: user.blogtype, auth_url: this.format_authorization_url(data)};
  _$jscoverage['tbase_oauth_v2.js'][66]++;
  process.nextTick((function () {
  _$jscoverage['tbase_oauth_v2.js'][67]++;
  callback(null, info);
}));
  _$jscoverage['tbase_oauth_v2.js'][69]++;
  return this;
});
_$jscoverage['tbase_oauth_v2.js'][72]++;
TBaseOauthV2.prototype.get_access_token = (function (user, callback) {
  _$jscoverage['tbase_oauth_v2.js'][73]++;
  var params = {type: "POST", user: user, playload: "string", api_host: this.config.oauth_host, request_method: "get_access_token"};
  _$jscoverage['tbase_oauth_v2.js'][80]++;
  var data = this.convert_token(user);
  _$jscoverage['tbase_oauth_v2.js'][81]++;
  data.grant_type = "authorization_code";
  _$jscoverage['tbase_oauth_v2.js'][82]++;
  data.client_secret = this.config.secret;
  _$jscoverage['tbase_oauth_v2.js'][83]++;
  var code = user.code || user.oauth_verifier || user.oauth_pin;
  _$jscoverage['tbase_oauth_v2.js'][84]++;
  if (code) {
    _$jscoverage['tbase_oauth_v2.js'][85]++;
    data.code = code;
  }
  _$jscoverage['tbase_oauth_v2.js'][88]++;
  params.data = data;
  _$jscoverage['tbase_oauth_v2.js'][89]++;
  var self = this;
  _$jscoverage['tbase_oauth_v2.js'][90]++;
  var url = self.config.oauth_access_token;
  _$jscoverage['tbase_oauth_v2.js'][91]++;
  self.send_request(url, params, (function (err, token) {
  _$jscoverage['tbase_oauth_v2.js'][92]++;
  if (err) {
    _$jscoverage['tbase_oauth_v2.js'][93]++;
    return callback(err);
  }
  _$jscoverage['tbase_oauth_v2.js'][99]++;
  token = self.format_access_token(token);
  _$jscoverage['tbase_oauth_v2.js'][100]++;
  if (! token.access_token) {
    _$jscoverage['tbase_oauth_v2.js'][101]++;
    var message = token.error || JSON.stringify(token);
    _$jscoverage['tbase_oauth_v2.js'][102]++;
    err = new Error(message);
    _$jscoverage['tbase_oauth_v2.js'][103]++;
    err.data = token;
    _$jscoverage['tbase_oauth_v2.js'][104]++;
    err.name = self.errorname("get_access_token");
    _$jscoverage['tbase_oauth_v2.js'][105]++;
    return callback(err);
  }
  _$jscoverage['tbase_oauth_v2.js'][107]++;
  token.blogtype = user.blogtype;
  _$jscoverage['tbase_oauth_v2.js'][108]++;
  callback(null, token);
}));
  _$jscoverage['tbase_oauth_v2.js'][110]++;
  return this;
});
_$jscoverage['tbase_oauth_v2.js'][113]++;
TBaseOauthV2.prototype.apply_auth = (function (url, args, user) {
  _$jscoverage['tbase_oauth_v2.js'][114]++;
  args.data = args.data || {};
  _$jscoverage['tbase_oauth_v2.js'][115]++;
  args.data.access_token = user.access_token;
});
_$jscoverage['tbase_oauth_v2.js'].source = ["/*!"," * node-weibo - lib/tbase_oauth_v2.js"," * Copyright(c) 2012 fengmk2 &lt;fengmk2@gmail.com&gt;"," * MIT Licensed"," */","","\"use strict\";","","/**"," * Module dependencies."," */","","var TBase = require('./tbase');","var inherits = require('util').inherits;","var utils = require('./utils');","var querystring = require('querystring');","","/**"," * TAPI Base class, support OAuth v2.0"," */","function TBaseOauthV2() {","  TBaseOauthV2.super_.call(this);","  this.config.oauth_version = '2.0';","}","","inherits(TBaseOauthV2, TBase);","module.exports = TBaseOauthV2;","","/**"," * Result formatters"," */","","TBaseOauthV2.prototype.format_access_token = function (token) {","  token = JSON.parse(token);","  return token;","};","","/**"," * OAuth"," */","","TBaseOauthV2.prototype.convert_token = function (user) {","  var params = {","    redirect_uri: user.oauth_callback || this.config.oauth_callback,","    client_id: this.config.appkey,","    response_type: 'code',","  };","  var oauth_scope = user.oauth_scope || this.config.oauth_scope;","  if (oauth_scope) {","    params.oauth_scope = oauth_scope;","  }","  if (user.state) {","    // An unguessable random string. It is used to protect against cross-site request forgery attacks.","    params.state = user.state;","  }","  return params;","};","","TBaseOauthV2.prototype.get_authorization_url = function (user, callback) {","  var data = this.convert_token(user);","  data.response_type = 'code';","  var info = {","    blogtype: user.blogtype,","    auth_url: this.format_authorization_url(data)","  };","  process.nextTick(function () {","    callback(null, info);","  });","  return this;","};","","TBaseOauthV2.prototype.get_access_token = function (user, callback) {","  var params = {","    type: 'POST',","    user: user,","    playload: 'string',","    api_host: this.config.oauth_host,","    request_method: 'get_access_token'","  };","  var data = this.convert_token(user);","  data.grant_type = 'authorization_code';","  data.client_secret = this.config.secret;","  var code = user.code || user.oauth_verifier || user.oauth_pin;","  if (code) {","    data.code = code;","  }","","  params.data = data;","  var self = this;","  var url = self.config.oauth_access_token;","  self.send_request(url, params, function (err, token) {","    if (err) {","      return callback(err);","    }","    // { access_token: '2.00EkofzBtMpzNBb9bc3108d8MwDTTE',","    // remind_in: '633971',","    // expires_in: 633971,","    // uid: '1827455832' }","    token = self.format_access_token(token);","    if (!token.access_token) {","      var message = token.error || JSON.stringify(token);","      err = new Error(message);","      err.data = token;","      err.name = self.errorname('get_access_token');","      return callback(err);","    }","    token.blogtype = user.blogtype;","    callback(null, token);","  });","  return this;","};","","TBaseOauthV2.prototype.apply_auth = function (url, args, user) {","  args.data = args.data || {};","  args.data.access_token = user.access_token;","};",""];
