// weibo/lib-cov/tqq.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['tqq.js']) {
  _$jscoverage['tqq.js'] = [];
  _$jscoverage['tqq.js'][7] = 0;
  _$jscoverage['tqq.js'][13] = 0;
  _$jscoverage['tqq.js'][14] = 0;
  _$jscoverage['tqq.js'][15] = 0;
  _$jscoverage['tqq.js'][17] = 0;
  _$jscoverage['tqq.js'][18] = 0;
  _$jscoverage['tqq.js'][20] = 0;
  _$jscoverage['tqq.js'][92] = 0;
  _$jscoverage['tqq.js'][95] = 0;
  _$jscoverage['tqq.js'][96] = 0;
  _$jscoverage['tqq.js'][102] = 0;
  _$jscoverage['tqq.js'][103] = 0;
  _$jscoverage['tqq.js'][104] = 0;
  _$jscoverage['tqq.js'][105] = 0;
  _$jscoverage['tqq.js'][106] = 0;
  _$jscoverage['tqq.js'][107] = 0;
  _$jscoverage['tqq.js'][108] = 0;
  _$jscoverage['tqq.js'][109] = 0;
  _$jscoverage['tqq.js'][110] = 0;
  _$jscoverage['tqq.js'][112] = 0;
  _$jscoverage['tqq.js'][113] = 0;
  _$jscoverage['tqq.js'][114] = 0;
  _$jscoverage['tqq.js'][115] = 0;
  _$jscoverage['tqq.js'][117] = 0;
  _$jscoverage['tqq.js'][120] = 0;
  _$jscoverage['tqq.js'][121] = 0;
  _$jscoverage['tqq.js'][128] = 0;
  _$jscoverage['tqq.js'][129] = 0;
  _$jscoverage['tqq.js'][130] = 0;
  _$jscoverage['tqq.js'][131] = 0;
  _$jscoverage['tqq.js'][132] = 0;
  _$jscoverage['tqq.js'][133] = 0;
  _$jscoverage['tqq.js'][135] = 0;
  _$jscoverage['tqq.js'][137] = 0;
  _$jscoverage['tqq.js'][163] = 0;
  _$jscoverage['tqq.js'][164] = 0;
  _$jscoverage['tqq.js'][165] = 0;
  _$jscoverage['tqq.js'][166] = 0;
  _$jscoverage['tqq.js'][167] = 0;
  _$jscoverage['tqq.js'][169] = 0;
  _$jscoverage['tqq.js'][172] = 0;
  _$jscoverage['tqq.js'][173] = 0;
  _$jscoverage['tqq.js'][216] = 0;
  _$jscoverage['tqq.js'][217] = 0;
  _$jscoverage['tqq.js'][218] = 0;
  _$jscoverage['tqq.js'][219] = 0;
  _$jscoverage['tqq.js'][220] = 0;
  _$jscoverage['tqq.js'][221] = 0;
  _$jscoverage['tqq.js'][223] = 0;
  _$jscoverage['tqq.js'][224] = 0;
  _$jscoverage['tqq.js'][225] = 0;
  _$jscoverage['tqq.js'][227] = 0;
  _$jscoverage['tqq.js'][228] = 0;
  _$jscoverage['tqq.js'][229] = 0;
  _$jscoverage['tqq.js'][230] = 0;
  _$jscoverage['tqq.js'][231] = 0;
  _$jscoverage['tqq.js'][233] = 0;
  _$jscoverage['tqq.js'][234] = 0;
  _$jscoverage['tqq.js'][236] = 0;
  _$jscoverage['tqq.js'][237] = 0;
  _$jscoverage['tqq.js'][239] = 0;
  _$jscoverage['tqq.js'][240] = 0;
  _$jscoverage['tqq.js'][241] = 0;
  _$jscoverage['tqq.js'][242] = 0;
  _$jscoverage['tqq.js'][244] = 0;
  _$jscoverage['tqq.js'][373] = 0;
  _$jscoverage['tqq.js'][374] = 0;
  _$jscoverage['tqq.js'][375] = 0;
  _$jscoverage['tqq.js'][376] = 0;
  _$jscoverage['tqq.js'][377] = 0;
  _$jscoverage['tqq.js'][378] = 0;
  _$jscoverage['tqq.js'][379] = 0;
  _$jscoverage['tqq.js'][380] = 0;
  _$jscoverage['tqq.js'][382] = 0;
  _$jscoverage['tqq.js'][383] = 0;
  _$jscoverage['tqq.js'][384] = 0;
  _$jscoverage['tqq.js'][386] = 0;
  _$jscoverage['tqq.js'][387] = 0;
  _$jscoverage['tqq.js'][389] = 0;
  _$jscoverage['tqq.js'][390] = 0;
  _$jscoverage['tqq.js'][391] = 0;
  _$jscoverage['tqq.js'][392] = 0;
  _$jscoverage['tqq.js'][393] = 0;
  _$jscoverage['tqq.js'][394] = 0;
  _$jscoverage['tqq.js'][395] = 0;
  _$jscoverage['tqq.js'][397] = 0;
  _$jscoverage['tqq.js'][398] = 0;
  _$jscoverage['tqq.js'][400] = 0;
  _$jscoverage['tqq.js'][402] = 0;
  _$jscoverage['tqq.js'][404] = 0;
  _$jscoverage['tqq.js'][406] = 0;
  _$jscoverage['tqq.js'][408] = 0;
  _$jscoverage['tqq.js'][410] = 0;
  _$jscoverage['tqq.js'][411] = 0;
  _$jscoverage['tqq.js'][414] = 0;
  _$jscoverage['tqq.js'][415] = 0;
  _$jscoverage['tqq.js'][417] = 0;
  _$jscoverage['tqq.js'][420] = 0;
  _$jscoverage['tqq.js'][421] = 0;
  _$jscoverage['tqq.js'][424] = 0;
  _$jscoverage['tqq.js'][425] = 0;
  _$jscoverage['tqq.js'][432] = 0;
  _$jscoverage['tqq.js'][435] = 0;
  _$jscoverage['tqq.js'][436] = 0;
  _$jscoverage['tqq.js'][437] = 0;
  _$jscoverage['tqq.js'][438] = 0;
  _$jscoverage['tqq.js'][439] = 0;
  _$jscoverage['tqq.js'][441] = 0;
  _$jscoverage['tqq.js'][444] = 0;
  _$jscoverage['tqq.js'][445] = 0;
  _$jscoverage['tqq.js'][448] = 0;
  _$jscoverage['tqq.js'][449] = 0;
  _$jscoverage['tqq.js'][452] = 0;
  _$jscoverage['tqq.js'][453] = 0;
  _$jscoverage['tqq.js'][457] = 0;
  _$jscoverage['tqq.js'][464] = 0;
  _$jscoverage['tqq.js'][466] = 0;
  _$jscoverage['tqq.js'][470] = 0;
  _$jscoverage['tqq.js'][473] = 0;
  _$jscoverage['tqq.js'][475] = 0;
  _$jscoverage['tqq.js'][478] = 0;
  _$jscoverage['tqq.js'][479] = 0;
  _$jscoverage['tqq.js'][480] = 0;
  _$jscoverage['tqq.js'][482] = 0;
  _$jscoverage['tqq.js'][483] = 0;
  _$jscoverage['tqq.js'][485] = 0;
  _$jscoverage['tqq.js'][488] = 0;
  _$jscoverage['tqq.js'][489] = 0;
  _$jscoverage['tqq.js'][492] = 0;
  _$jscoverage['tqq.js'][495] = 0;
  _$jscoverage['tqq.js'][496] = 0;
  _$jscoverage['tqq.js'][517] = 0;
  _$jscoverage['tqq.js'][518] = 0;
  _$jscoverage['tqq.js'][520] = 0;
  _$jscoverage['tqq.js'][521] = 0;
  _$jscoverage['tqq.js'][522] = 0;
  _$jscoverage['tqq.js'][523] = 0;
  _$jscoverage['tqq.js'][525] = 0;
  _$jscoverage['tqq.js'][526] = 0;
  _$jscoverage['tqq.js'][527] = 0;
  _$jscoverage['tqq.js'][528] = 0;
  _$jscoverage['tqq.js'][531] = 0;
  _$jscoverage['tqq.js'][532] = 0;
  _$jscoverage['tqq.js'][536] = 0;
  _$jscoverage['tqq.js'][537] = 0;
  _$jscoverage['tqq.js'][538] = 0;
  _$jscoverage['tqq.js'][540] = 0;
  _$jscoverage['tqq.js'][541] = 0;
  _$jscoverage['tqq.js'][543] = 0;
  _$jscoverage['tqq.js'][550] = 0;
  _$jscoverage['tqq.js'][551] = 0;
  _$jscoverage['tqq.js'][552] = 0;
  _$jscoverage['tqq.js'][553] = 0;
  _$jscoverage['tqq.js'][555] = 0;
  _$jscoverage['tqq.js'][556] = 0;
  _$jscoverage['tqq.js'][557] = 0;
  _$jscoverage['tqq.js'][559] = 0;
  _$jscoverage['tqq.js'][561] = 0;
  _$jscoverage['tqq.js'][564] = 0;
  _$jscoverage['tqq.js'][565] = 0;
  _$jscoverage['tqq.js'][566] = 0;
  _$jscoverage['tqq.js'][567] = 0;
  _$jscoverage['tqq.js'][569] = 0;
  _$jscoverage['tqq.js'][571] = 0;
  _$jscoverage['tqq.js'][574] = 0;
  _$jscoverage['tqq.js'][575] = 0;
  _$jscoverage['tqq.js'][576] = 0;
  _$jscoverage['tqq.js'][579] = 0;
  _$jscoverage['tqq.js'][580] = 0;
  _$jscoverage['tqq.js'][581] = 0;
  _$jscoverage['tqq.js'][582] = 0;
  _$jscoverage['tqq.js'][584] = 0;
  _$jscoverage['tqq.js'][585] = 0;
  _$jscoverage['tqq.js'][586] = 0;
  _$jscoverage['tqq.js'][588] = 0;
  _$jscoverage['tqq.js'][595] = 0;
  _$jscoverage['tqq.js'][596] = 0;
  _$jscoverage['tqq.js'][597] = 0;
  _$jscoverage['tqq.js'][598] = 0;
  _$jscoverage['tqq.js'][600] = 0;
  _$jscoverage['tqq.js'][603] = 0;
  _$jscoverage['tqq.js'][604] = 0;
  _$jscoverage['tqq.js'][605] = 0;
  _$jscoverage['tqq.js'][606] = 0;
  _$jscoverage['tqq.js'][607] = 0;
  _$jscoverage['tqq.js'][608] = 0;
  _$jscoverage['tqq.js'][609] = 0;
  _$jscoverage['tqq.js'][611] = 0;
  _$jscoverage['tqq.js'][613] = 0;
  _$jscoverage['tqq.js'][616] = 0;
  _$jscoverage['tqq.js'][617] = 0;
  _$jscoverage['tqq.js'][624] = 0;
  _$jscoverage['tqq.js'][625] = 0;
  _$jscoverage['tqq.js'][626] = 0;
  _$jscoverage['tqq.js'][627] = 0;
  _$jscoverage['tqq.js'][628] = 0;
  _$jscoverage['tqq.js'][629] = 0;
  _$jscoverage['tqq.js'][631] = 0;
  _$jscoverage['tqq.js'][635] = 0;
}
_$jscoverage['tqq.js'][7]++;
"use strict";
_$jscoverage['tqq.js'][13]++;
var inherits = require("util").inherits;
_$jscoverage['tqq.js'][14]++;
var TBase = require("./tbase");
_$jscoverage['tqq.js'][15]++;
var utils = require("./utils");
_$jscoverage['tqq.js'][17]++;
function TQQAPI(options) {
  _$jscoverage['tqq.js'][18]++;
  TQQAPI.super_.call(this);
  _$jscoverage['tqq.js'][20]++;
  var config = utils.extend({}, options, {host: "http://open.t.qq.com/api", result_format: "", oauth_host: "https://open.t.qq.com", oauth_authorize: "/cgi-bin/authorize", oauth_request_token: "/cgi-bin/request_token", oauth_access_token: "/cgi-bin/access_token", oauth_params_by_get: true, support_comment: false, support_do_comment: true, support_repost_timeline: true, support_favorites_max_id: true, reply_dont_need_at_screen_name: true, rt_at_name: true, repost_delimiter: " || ", support_counts: false, home_timeline: "/statuses/home_timeline", mentions: "/statuses/mentions_timeline", comments_timeline: "/statuses/mentions_timeline", comments_mentions: "/statuses/mentions_timeline", repost_timeline: "/t/re_list", followers: "/friends/user_fanslist", friends: "/friends/user_idollist", favorites: "/fav/list_t", favorite_show: null, favorite_create: "/fav/addt", favorite_destroy: "/fav/delt", count: "/t/re_count", show: "/t/show", update: "/t/add", upload: "/t/add_pic", repost: "/t/re_add", comment_create: "/t/comment", comment_reply: "/t/comment", comments: "/t/re_list", destroy: "/t/del", destroy_msg: "/private/del", direct_messages: "/private/recv", sent_direct_messages: "/private/send", new_message: "/private/add", rate_limit_status: "/account/rate_limit_status", friendships_create: "/friends/add", friendships_destroy: "/friends/del", friendships_show: "/friends/check", reset_count: "/statuses/reset_count", user_show: "/user/other_info", tags: "/tags", create_tag: "/tags/create", destroy_tag: "/tags/destroy", tags_suggestions: "/tags/suggestions", search: "/search/t", user_search: "/search/user", verify_credentials: "/user/info", gender_map: {0: "n", 1: "m", 2: "f"}, support_comment_destroy: false, support_comments_mentions: false, support_comments_to_me: false, support_comments_by_me: false});
  _$jscoverage['tqq.js'][92]++;
  this.init(config);
}
_$jscoverage['tqq.js'][95]++;
inherits(TQQAPI, TBase);
_$jscoverage['tqq.js'][96]++;
module.exports = TQQAPI;
_$jscoverage['tqq.js'][102]++;
TQQAPI.prototype.detect_error = (function (method, res, playload, data) {
  _$jscoverage['tqq.js'][103]++;
  var headers = res.headers;
  _$jscoverage['tqq.js'][104]++;
  var err;
  _$jscoverage['tqq.js'][105]++;
  if (res.statusCode === 200 && headers.status) {
    _$jscoverage['tqq.js'][106]++;
    err = new Error(headers.status);
  }
  else {
    _$jscoverage['tqq.js'][107]++;
    if (data.errcode && data.msg) {
      _$jscoverage['tqq.js'][108]++;
      err = new Error(data.msg);
    }
    else {
      _$jscoverage['tqq.js'][109]++;
      if (! data.data && data.msg && data.msg !== "ok") {
        _$jscoverage['tqq.js'][110]++;
        err = new Error(data.msg);
      }
    }
  }
  _$jscoverage['tqq.js'][112]++;
  if (err) {
    _$jscoverage['tqq.js'][113]++;
    err.name = this.errorname(method);
    _$jscoverage['tqq.js'][114]++;
    err.data = data;
    _$jscoverage['tqq.js'][115]++;
    return err;
  }
  _$jscoverage['tqq.js'][117]++;
  return TQQAPI.super_.prototype.detect_error.call(this, method, res, playload, data);
});
_$jscoverage['tqq.js'][120]++;
TQQAPI.prototype.url_encode = (function (text) {
  _$jscoverage['tqq.js'][121]++;
  return text;
});
_$jscoverage['tqq.js'][128]++;
TQQAPI.prototype.get_result_items = (function (data, playload, args) {
  _$jscoverage['tqq.js'][129]++;
  if (playload === "count") {
    _$jscoverage['tqq.js'][130]++;
    var counts = [];
    _$jscoverage['tqq.js'][131]++;
    for (var id in data) {
      _$jscoverage['tqq.js'][132]++;
      var item = data[id];
      _$jscoverage['tqq.js'][133]++;
      counts.push({id: id, reposts: item.count, comments: item.mcount});
}
    _$jscoverage['tqq.js'][135]++;
    return counts;
  }
  _$jscoverage['tqq.js'][137]++;
  return data.info;
});
_$jscoverage['tqq.js'][163]++;
TQQAPI.prototype.format_result = (function (data, playload, args) {
  _$jscoverage['tqq.js'][164]++;
  data = data.data;
  _$jscoverage['tqq.js'][165]++;
  var result = TQQAPI.super_.prototype.format_result.call(this, data, playload, args);
  _$jscoverage['tqq.js'][166]++;
  if (data.user) {
    _$jscoverage['tqq.js'][167]++;
    result.users = data.user;
  }
  _$jscoverage['tqq.js'][169]++;
  return result;
});
_$jscoverage['tqq.js'][172]++;
TQQAPI.prototype.format_search_status = (function (status, args) {
  _$jscoverage['tqq.js'][173]++;
  throw new Error("Must override this method.");
});
_$jscoverage['tqq.js'][216]++;
TQQAPI.prototype.format_status = (function (data, args) {
  _$jscoverage['tqq.js'][217]++;
  var status = {};
  _$jscoverage['tqq.js'][218]++;
  status.id = String(data.id);
  _$jscoverage['tqq.js'][219]++;
  status.t_url = "http://t.qq.com/p/t/" + data.id;
  _$jscoverage['tqq.js'][220]++;
  if (! data.timestamp) {
    _$jscoverage['tqq.js'][221]++;
    return status;
  }
  _$jscoverage['tqq.js'][223]++;
  status.created_at = new Date(data.timestamp * 1000);
  _$jscoverage['tqq.js'][224]++;
  status.text = data.origtext;
  _$jscoverage['tqq.js'][225]++;
  status.source = "<a href=\"" + data.fromurl + "\">" + data.from + "</a>";
  _$jscoverage['tqq.js'][227]++;
  if (data.image && data.image[0]) {
    _$jscoverage['tqq.js'][228]++;
    var image = data.image[0];
    _$jscoverage['tqq.js'][229]++;
    status.thumbnail_pic = image + "/160";
    _$jscoverage['tqq.js'][230]++;
    status.bmiddle_pic = image + "/460";
    _$jscoverage['tqq.js'][231]++;
    status.original_pic = image + "/2000";
  }
  _$jscoverage['tqq.js'][233]++;
  if (data.latitude && String(data.latitude) !== "0") {
    _$jscoverage['tqq.js'][234]++;
    status.geo = this.format_geo(data, args);
  }
  _$jscoverage['tqq.js'][236]++;
  if (data.name) {
    _$jscoverage['tqq.js'][237]++;
    status.user = this.format_user(data, args);
  }
  _$jscoverage['tqq.js'][239]++;
  status.reposts_count = data.count || 0;
  _$jscoverage['tqq.js'][240]++;
  status.comments_count = data.mcount || 0;
  _$jscoverage['tqq.js'][241]++;
  if (data.source) {
    _$jscoverage['tqq.js'][242]++;
    status.retweeted_status = this.format_status(data.source, args);
  }
  _$jscoverage['tqq.js'][244]++;
  return status;
});
_$jscoverage['tqq.js'][373]++;
TQQAPI.prototype.format_user = (function (data, args) {
  _$jscoverage['tqq.js'][374]++;
  var user = {};
  _$jscoverage['tqq.js'][375]++;
  user.id = data.name;
  _$jscoverage['tqq.js'][376]++;
  user.t_url = "http://t.qq.com/" + data.name;
  _$jscoverage['tqq.js'][377]++;
  user.screen_name = data.nick;
  _$jscoverage['tqq.js'][378]++;
  user.name = data.name;
  _$jscoverage['tqq.js'][379]++;
  user.location = data.location || "";
  _$jscoverage['tqq.js'][380]++;
  user.description = data.introduction || "";
  _$jscoverage['tqq.js'][382]++;
  if (data.head) {
    _$jscoverage['tqq.js'][383]++;
    user.profile_image_url = data.head + "/50";
    _$jscoverage['tqq.js'][384]++;
    user.avatar_large = data.head + "/180";
  }
  else {
    _$jscoverage['tqq.js'][386]++;
    user.profile_image_url = "http://mat1.gtimg.com/www/mb/images/head_50.jpg";
    _$jscoverage['tqq.js'][387]++;
    user.avatar_large = "http://mat1.gtimg.com/www/mb/images/head_180.jpg";
  }
  _$jscoverage['tqq.js'][389]++;
  user.gender = this.config.gender_map[data.sex || 0];
  _$jscoverage['tqq.js'][390]++;
  user.followers_count = data.fansnum || 0;
  _$jscoverage['tqq.js'][391]++;
  user.friends_count = data.idolnum || 0;
  _$jscoverage['tqq.js'][392]++;
  user.statuses_count = data.tweetnum || 0;
  _$jscoverage['tqq.js'][393]++;
  user.favourites_count = data.favnum || 0;
  _$jscoverage['tqq.js'][394]++;
  if (data.regtime) {
    _$jscoverage['tqq.js'][395]++;
    user.created_at = new Date(data.regtime * 1000);
  }
  _$jscoverage['tqq.js'][397]++;
  user.following = data.ismyidol || false;
  _$jscoverage['tqq.js'][398]++;
  user.follow_me = data.ismyfans || false;
  _$jscoverage['tqq.js'][400]++;
  user.allow_all_act_msg = data.send_private_flag === 2;
  _$jscoverage['tqq.js'][402]++;
  user.verified = ! ! data.isvip;
  _$jscoverage['tqq.js'][404]++;
  user.verified_reason = data.verifyinfo || "";
  _$jscoverage['tqq.js'][406]++;
  user.allow_all_comment = true;
  _$jscoverage['tqq.js'][408]++;
  user.bi_followers_count = data.mutual_fans_num || 0;
  _$jscoverage['tqq.js'][410]++;
  if (data.tweetinfo && data.tweetinfo[0]) {
    _$jscoverage['tqq.js'][411]++;
    user.status = this.format_status(data.tweetinfo[0], args);
  }
  _$jscoverage['tqq.js'][414]++;
  if (data.tag) {
    _$jscoverage['tqq.js'][415]++;
    user.tags = data.tag;
  }
  _$jscoverage['tqq.js'][417]++;
  return user;
});
_$jscoverage['tqq.js'][420]++;
TQQAPI.prototype.format_count = (function (count, args) {
  _$jscoverage['tqq.js'][421]++;
  return count;
});
_$jscoverage['tqq.js'][424]++;
TQQAPI.prototype.format_geo = (function (data, args) {
  _$jscoverage['tqq.js'][425]++;
  var geo = {longitude: data.longitude, latitude: data.latitude, address: data.geo};
  _$jscoverage['tqq.js'][432]++;
  return geo;
});
_$jscoverage['tqq.js'][435]++;
TQQAPI.prototype.format_comment = (function (data, args) {
  _$jscoverage['tqq.js'][436]++;
  var comment = this.format_status(data, args);
  _$jscoverage['tqq.js'][437]++;
  if (comment.retweeted_status) {
    _$jscoverage['tqq.js'][438]++;
    comment.status = comment.retweeted_status;
    _$jscoverage['tqq.js'][439]++;
    delete comment.retweeted_status;
  }
  _$jscoverage['tqq.js'][441]++;
  return comment;
});
_$jscoverage['tqq.js'][444]++;
TQQAPI.prototype.format_message = (function (message, args) {
  _$jscoverage['tqq.js'][445]++;
  throw new Error("Must override this method.");
});
_$jscoverage['tqq.js'][448]++;
TQQAPI.prototype.format_emotion = (function (emotion, args) {
  _$jscoverage['tqq.js'][449]++;
  throw new Error("Must override this method.");
});
_$jscoverage['tqq.js'][452]++;
TQQAPI.prototype.format_favorite = (function (status, args) {
  _$jscoverage['tqq.js'][453]++;
  var favorite = {created_at: new Date(status.storetime * 1000), status: this.format_status(status)};
  _$jscoverage['tqq.js'][457]++;
  return favorite;
});
_$jscoverage['tqq.js'][464]++;
TQQAPI.prototype.convert_comment = (function (comment) {
  _$jscoverage['tqq.js'][466]++;
  var data = {content: comment.comment, reid: comment.id};
  _$jscoverage['tqq.js'][470]++;
  return data;
});
_$jscoverage['tqq.js'][473]++;
TQQAPI.prototype.convert_status = (function (status) {
  _$jscoverage['tqq.js'][475]++;
  var data = {content: status.status};
  _$jscoverage['tqq.js'][478]++;
  if (status["long"]) {
    _$jscoverage['tqq.js'][479]++;
    data.longitude = status["long"];
    _$jscoverage['tqq.js'][480]++;
    data.latitude = status.lat;
  }
  _$jscoverage['tqq.js'][482]++;
  if (status.id) {
    _$jscoverage['tqq.js'][483]++;
    data.reid = status.id;
  }
  _$jscoverage['tqq.js'][485]++;
  return data;
});
_$jscoverage['tqq.js'][488]++;
TQQAPI.prototype.convert_user = (function (user) {
  _$jscoverage['tqq.js'][489]++;
  var data = {name: user.uid || user.screen_name};
  _$jscoverage['tqq.js'][492]++;
  return data;
});
_$jscoverage['tqq.js'][495]++;
TQQAPI.prototype.convert_ids = (function (ids) {
  _$jscoverage['tqq.js'][496]++;
  return ({ids: ids, flag: "2"});
});
_$jscoverage['tqq.js'][517]++;
TQQAPI.prototype.convert_cursor = (function (cursor) {
  _$jscoverage['tqq.js'][518]++;
  var data = {};
  _$jscoverage['tqq.js'][520]++;
  data.type = String(1 | 2 | 8 | 16 | 32);
  _$jscoverage['tqq.js'][521]++;
  data.contenttype = "0";
  _$jscoverage['tqq.js'][522]++;
  data.reqnum = cursor.count;
  _$jscoverage['tqq.js'][523]++;
  if (cursor.max_id) {
    _$jscoverage['tqq.js'][525]++;
    data.pageflag = "1";
    _$jscoverage['tqq.js'][526]++;
    data.pagetime = cursor.max_time;
    _$jscoverage['tqq.js'][527]++;
    data.lastid = cursor.max_id;
  }
  else {
    _$jscoverage['tqq.js'][528]++;
    if (cursor.since_id) {
      _$jscoverage['tqq.js'][531]++;
      data.pageflag = "2";
      _$jscoverage['tqq.js'][532]++;
      data.pagetime = cursor.since_time;
    }
    else {
      _$jscoverage['tqq.js'][536]++;
      data.pageflag = "0";
      _$jscoverage['tqq.js'][537]++;
      data.pagetime = "0";
      _$jscoverage['tqq.js'][538]++;
      data.lastid = "0";
    }
  }
  _$jscoverage['tqq.js'][540]++;
  if (typeof cursor.callback === "function") {
    _$jscoverage['tqq.js'][541]++;
    data = cursor.callback(data);
  }
  _$jscoverage['tqq.js'][543]++;
  return data;
});
_$jscoverage['tqq.js'][550]++;
TQQAPI.prototype.repost_timeline = (function (user, cursor, callback) {
  _$jscoverage['tqq.js'][551]++;
  cursor.callback = (function (data) {
  _$jscoverage['tqq.js'][552]++;
  data.rootid = cursor.id;
  _$jscoverage['tqq.js'][553]++;
  data.flag = "0";
  _$jscoverage['tqq.js'][555]++;
  if (data.lastid) {
    _$jscoverage['tqq.js'][556]++;
    data.twitterid = data.lastid;
    _$jscoverage['tqq.js'][557]++;
    delete data.lastid;
  }
  _$jscoverage['tqq.js'][559]++;
  return data;
});
  _$jscoverage['tqq.js'][561]++;
  return TQQAPI.super_.prototype.repost_timeline.call(this, user, cursor, callback);
});
_$jscoverage['tqq.js'][564]++;
TQQAPI.prototype.user_timeline = (function (user, cursor, callback) {
  _$jscoverage['tqq.js'][565]++;
  cursor.callback = (function (data) {
  _$jscoverage['tqq.js'][566]++;
  if (cursor.uid || cursor.screen_name) {
    _$jscoverage['tqq.js'][567]++;
    data.name = cursor.uid || cursor.screen_name;
  }
  _$jscoverage['tqq.js'][569]++;
  return data;
});
  _$jscoverage['tqq.js'][571]++;
  return TQQAPI.super_.prototype.user_timeline.call(this, user, cursor, callback);
});
_$jscoverage['tqq.js'][574]++;
TQQAPI.prototype.search = (function (user, query, cursor, callback) {
  _$jscoverage['tqq.js'][575]++;
  cursor = cursor || {};
  _$jscoverage['tqq.js'][576]++;
  var q = {keyword: query.q};
  _$jscoverage['tqq.js'][579]++;
  if (query["long"] && query.lat && query.radius) {
    _$jscoverage['tqq.js'][580]++;
    q.longitude = query["long"];
    _$jscoverage['tqq.js'][581]++;
    q.latitude = query.lat;
    _$jscoverage['tqq.js'][582]++;
    q.radius = query.radius;
  }
  _$jscoverage['tqq.js'][584]++;
  cursor.callback = (function (data) {
  _$jscoverage['tqq.js'][585]++;
  data.pagesize = data.reqnum || 20;
  _$jscoverage['tqq.js'][586]++;
  return data;
});
  _$jscoverage['tqq.js'][588]++;
  return TQQAPI.super_.prototype.search.call(this, user, q, cursor, callback);
});
_$jscoverage['tqq.js'][595]++;
TQQAPI.prototype.comments_timeline = (function (user, cursor, callback) {
  _$jscoverage['tqq.js'][596]++;
  cursor.callback = (function (data) {
  _$jscoverage['tqq.js'][597]++;
  data.type = String(64);
  _$jscoverage['tqq.js'][598]++;
  return data;
});
  _$jscoverage['tqq.js'][600]++;
  return TQQAPI.super_.prototype.comments_timeline.call(this, user, cursor, callback);
});
_$jscoverage['tqq.js'][603]++;
TQQAPI.prototype.comments = (function (user, cursor, callback) {
  _$jscoverage['tqq.js'][604]++;
  cursor.callback = (function (data) {
  _$jscoverage['tqq.js'][605]++;
  data.rootid = cursor.id;
  _$jscoverage['tqq.js'][606]++;
  data.flag = "1";
  _$jscoverage['tqq.js'][607]++;
  if (data.lastid) {
    _$jscoverage['tqq.js'][608]++;
    data.twitterid = data.lastid;
    _$jscoverage['tqq.js'][609]++;
    delete data.lastid;
  }
  _$jscoverage['tqq.js'][611]++;
  return data;
});
  _$jscoverage['tqq.js'][613]++;
  return TQQAPI.super_.prototype.comments.call(this, user, cursor, callback);
});
_$jscoverage['tqq.js'][616]++;
TQQAPI.prototype.comment_destroy = (function (user, cid, callback) {
  _$jscoverage['tqq.js'][617]++;
  callback(new TypeError("comment_destroy not support."));
});
_$jscoverage['tqq.js'][624]++;
TQQAPI.prototype.favorite_show = (function (user, id, callback) {
  _$jscoverage['tqq.js'][625]++;
  var self = this;
  _$jscoverage['tqq.js'][626]++;
  self.show(user, id, (function (err, status) {
  _$jscoverage['tqq.js'][627]++;
  if (err) {
    _$jscoverage['tqq.js'][628]++;
    err.name = self.errorname("favorite_show");
    _$jscoverage['tqq.js'][629]++;
    return callback(err);
  }
  _$jscoverage['tqq.js'][631]++;
  var favorite = {status: status, created_at: status.created_at};
  _$jscoverage['tqq.js'][635]++;
  callback(null, favorite);
}));
});
_$jscoverage['tqq.js'].source = ["/*!"," * node-weibo - lib/tqq.js"," * Copyright(c) 2012 fengmk2 &lt;fengmk2@gmail.com&gt;"," * MIT Licensed"," */","","\"use strict\";","","/**"," * Module dependencies."," */","","var inherits = require('util').inherits;","var TBase = require('./tbase');","var utils = require('./utils');","","function TQQAPI(options) {","  TQQAPI.super_.call(this);","","  var config = utils.extend({}, options, {","    host: 'http://open.t.qq.com/api',","    result_format: '',","    oauth_host: 'https://open.t.qq.com',","    oauth_authorize:      '/cgi-bin/authorize',","    oauth_request_token:  '/cgi-bin/request_token',","    oauth_access_token:   '/cgi-bin/access_token',","","    // &#231;&#171;&#159;&#231;&#132;&#182;&#230;&#152;&#175;&#233;&#128;&#154;&#232;&#191;&#135;get&#228;&#188;&#160;&#233;&#128;&#146;","    oauth_params_by_get: true,","    support_comment: false, // &#228;&#184;&#141;&#230;&#148;&#175;&#230;&#140;&#129;comment_timeline","    support_do_comment: true,","    support_repost_timeline: true, // &#230;&#148;&#175;&#230;&#140;&#129;&#230;&#159;&#165;&#231;&#156;&#139;&#232;&#189;&#172;&#229;&#143;&#145;&#229;&#136;&#151;&#232;&#161;&#168;","    support_favorites_max_id: true,","    reply_dont_need_at_screen_name: true, // @&#229;&#155;&#158;&#229;&#164;&#141;&#230;&#159;&#144;&#230;&#157;&#161;&#229;&#190;&#174;&#229;&#141;&#154; &#230;&#151;&#160;&#233;&#156;&#128;&#229;&#161;&#171;&#229;&#133;&#133;@screen_name ","    rt_at_name: true, // RT&#231;&#154;&#132;@name&#232;&#128;&#140;&#228;&#184;&#141;&#230;&#152;&#175;@screen_name","    repost_delimiter: ' || ', //&#232;&#189;&#172;&#229;&#143;&#145;&#230;&#151;&#182;&#231;&#154;&#132;&#229;&#136;&#134;&#233;&#154;&#148;&#231;&#172;&#166;","    support_counts: false, // &#229;&#143;&#170;&#230;&#156;&#137;rt_count&#232;&#191;&#153;&#228;&#184;&#170;&#239;&#188;&#140;&#228;&#184;&#141;&#232;&#191;&#135;&#232;&#178;&#140;&#228;&#188;&#188;&#230;&#156;&#137;&#233;&#151;&#174;&#233;&#162;&#152;&#239;&#188;&#140;&#230;&#128;&#187;&#230;&#152;&#175;404&#227;&#128;&#130;&#230;&#154;&#130;&#230;&#151;&#182;&#233;&#154;&#144;&#232;&#151;&#143;","","    home_timeline:        '/statuses/home_timeline',","    mentions:             '/statuses/mentions_timeline',","    comments_timeline:    '/statuses/mentions_timeline',","    comments_mentions:    '/statuses/mentions_timeline',","    ","    repost_timeline:      '/t/re_list',","","    followers:            '/friends/user_fanslist',","    friends:              '/friends/user_idollist',","    favorites:            '/fav/list_t',","    favorite_show:        null, // use show() to mock this api","    favorite_create:      '/fav/addt',","    favorite_destroy:     '/fav/delt',","    count:                '/t/re_count', //&#228;&#187;&#133;&#228;&#187;&#133;&#230;&#152;&#175;&#232;&#189;&#172;&#230;&#146;&#173;&#230;&#149;&#176;","    show:                 '/t/show',","    update:               '/t/add',","    upload:               '/t/add_pic',","    repost:               '/t/re_add',","    comment_create:       '/t/comment',","    comment_reply:        '/t/comment',","    comments:             '/t/re_list',","    destroy:              '/t/del',","    destroy_msg:          '/private/del',","    direct_messages:      '/private/recv',","    sent_direct_messages: '/private/send',","    new_message:          '/private/add',","    rate_limit_status:    '/account/rate_limit_status',","    friendships_create:   '/friends/add',","    friendships_destroy:  '/friends/del',","    friendships_show:     '/friends/check',","    reset_count:          '/statuses/reset_count',","    user_show:            '/user/other_info',","    ","    // &#231;&#148;&#168;&#230;&#136;&#183;&#230;&#160;&#135;&#231;&#173;&#190;","    tags:                 '/tags',","    create_tag:           '/tags/create',","    destroy_tag:          '/tags/destroy',","    tags_suggestions:     '/tags/suggestions',","    ","    // &#230;&#144;&#156;&#231;&#180;&#162;","    search:               '/search/t',","    user_search:          '/search/user',","    verify_credentials:   '/user/info',","    ","    gender_map: {0: 'n', 1: 'm', 2: 'f'},","","    // support apis","    support_comment_destroy: false,","    support_comments_mentions: false,","    support_comments_to_me: false,","    support_comments_by_me: false,","  });","","  this.init(config);","}","","inherits(TQQAPI, TBase);","module.exports = TQQAPI;","","/**"," * Utils methods"," */","","TQQAPI.prototype.detect_error = function (method, res, playload, data) {","  var headers = res.headers;","  var err;","  if (res.statusCode === 200 &amp;&amp; headers.status) {","    err = new Error(headers.status);","  } else if (data.errcode &amp;&amp; data.msg) {","    err = new Error(data.msg);","  } else if (!data.data &amp;&amp; data.msg &amp;&amp; data.msg !== 'ok') {","    err = new Error(data.msg);","  }","  if (err) {","    err.name = this.errorname(method);","    err.data = data;","    return err;","  }","  return TQQAPI.super_.prototype.detect_error.call(this, method, res, playload, data);","};","","TQQAPI.prototype.url_encode = function (text) {","  return text;","};","","/**"," * Result getters"," */","","TQQAPI.prototype.get_result_items = function (data, playload, args) {","  if (playload === 'count') {","    var counts = [];","    for (var id in data) {","      var item = data[id];","      counts.push({id: id, reposts: item.count, comments: item.mcount});","    }","    return counts;","  }","  return data.info;","};","","/**"," * { hasnext: 0,","     info: ","      [ [Object],","        [Object],","        [Object],","        [Object],","        [Object],","        [Object],","        [Object],","        [Object],","        [Object] ],","     timestamp: 1348753615,","     user: { GreenMango: '&#233;&#157;&#146;&#232;&#138;&#146;', 'node-weibo': 'node-weibo' } },"," */","// TQQAPI.prototype.get_pagging_cursor = function (data, playload, args) {","//   return {};","// };","","/**"," * Result formatters"," */","","TQQAPI.prototype.format_result = function (data, playload, args) {","  data = data.data;","  var result = TQQAPI.super_.prototype.format_result.call(this, data, playload, args);","  if (data.user) {","    result.users = data.user;","  }","  return result;","};","","TQQAPI.prototype.format_search_status = function (status, args) {","  throw new Error('Must override this method.');","};","","/**"," *","{ city_code: '1',","  count: 0,","  country_code: '1',","  emotiontype: 0,","  emotionurl: '',","  from: '&#229;&#190;&#174;&#229;&#141;&#154;&#229;&#188;&#128;&#230;&#148;&#190;&#229;&#185;&#179;&#229;&#143;&#176;',","  fromurl: 'http://wiki.open.t.qq.com/index.php/%E4%BA%A7%E5%93%81%E7%B1%BBFAQ#.E6.8F.90.E4.BA.A4.E5.BA.94.E7.94.A8.E6.9D.A5.E6.BA.90.E5.AD.97.E6.AE.B5.E5.AE.A1.E6.A0.B8.E8.83.BD.E5.BE.97.E5.88.B0.E4.BB.80.E4.B9.88.E5.A5.BD.E5.A4.84.EF.BC.9F\\n',","  geo: '&#229;&#185;&#191;&#228;&#184;&#156;&#231;&#156;&#129;&#228;&#184;&#173;&#229;&#177;&#177;&#229;&#184;&#130;&#229;&#186;&#183;&#228;&#185;&#144;&#232;&#183;&#175;&#239;&#188;&#145;&#239;&#188;&#144;&#229;&#143;&#183;',","  head: 'http://app.qlogo.cn/mbloghead/cb1c4eb21aa2b52a233a',","  id: '102460077174373',","  image: null,","  isrealname: 2,","  isvip: 0,","  jing: '113.421234',","  latitude: '22.354231',","  location: '&#228;&#184;&#173;&#229;&#155;&#189; &#230;&#181;&#153;&#230;&#177;&#159; &#230;&#157;&#173;&#229;&#183;&#158;',","  longitude: '113.421234',","  mcount: 0,","  music: null,","  name: 'node-weibo',","  nick: 'node-weibo',","  openid: 'EA68676D5E9DA465822CD0CEB2DC6EF5',","  origtext: '&#232;&#191;&#153;&#230;&#152;&#175;update(user, status, callback) &#231;&#154;&#132;&#229;&#141;&#149;&#229;&#133;&#131;&#230;&#181;&#139;&#232;&#175;&#149;&#239;&#188;&#140;&#229;&#189;&#147;&#229;&#137;&#141;&#230;&#151;&#182;&#233;&#151;&#180; Thu Sep 27 2012 17:04:25 GMT+0800 (CST)',","  province_code: '33',","  self: 1,","  source: null,","  status: 0,","  text: '&#232;&#191;&#153;&#230;&#152;&#175;update(user, status, callback) &#231;&#154;&#132;&#229;&#141;&#149;&#229;&#133;&#131;&#230;&#181;&#139;&#232;&#175;&#149;&#239;&#188;&#140;&#229;&#189;&#147;&#229;&#137;&#141;&#230;&#151;&#182;&#233;&#151;&#180; Thu Sep 27 2012 17:04:25 GMT+0800 (CST)',","  timestamp: 1348736665,","  type: 1,","  user: { 'node-weibo': 'node-weibo' },","  video: null,","  wei: '22.354231' }",""," * @param {[type]} status [description]"," * @param {[type]} args [description]"," * @return {[type]} [description]"," */","TQQAPI.prototype.format_status = function (data, args) {","  var status = {};","  status.id = String(data.id);","  status.t_url = 'http://t.qq.com/p/t/' + data.id;","  if (!data.timestamp) {","    return status;","  }","  status.created_at = new Date(data.timestamp * 1000);","  status.text = data.origtext;","  status.source = '&lt;a href=\"' + data.fromurl + '\"&gt;' + data.from + '&lt;/a&gt;';","  // status.favorited = ","  if (data.image &amp;&amp; data.image[0]) {","    var image = data.image[0];","    status.thumbnail_pic = image + '/160';","    status.bmiddle_pic = image + '/460';","    status.original_pic = image + '/2000';","  }","  if (data.latitude &amp;&amp; String(data.latitude) !== '0') {","    status.geo = this.format_geo(data, args);","  }","  if (data.name) {","    status.user = this.format_user(data, args);","  }","  status.reposts_count = data.count || 0;","  status.comments_count = data.mcount || 0;","  if (data.source) {","    status.retweeted_status = this.format_status(data.source, args);","  }","  return status;","  // type:&#229;&#190;&#174;&#229;&#141;&#154;&#231;&#177;&#187;&#229;&#158;&#139; 1-&#229;&#142;&#159;&#229;&#136;&#155;&#229;&#143;&#145;&#232;&#161;&#168;&#227;&#128;&#129;2-&#232;&#189;&#172;&#232;&#189;&#189;&#227;&#128;&#129;3-&#231;&#167;&#129;&#228;&#191;&#161; 4-&#229;&#155;&#158;&#229;&#164;&#141; 5-&#231;&#169;&#186;&#229;&#155;&#158; 6-&#230;&#143;&#144;&#229;&#143;&#138; 7: &#231;&#130;&#185;&#232;&#175;&#132;","//      status.status_type = data.type;","//       if(data.type == 7) {","//         // &#232;&#133;&#190;&#232;&#174;&#175;&#231;&#154;&#132;&#231;&#130;&#185;&#232;&#175;&#132;&#228;&#188;&#154;&#228;&#187;&#138;&#230;&#151;&#165;hometimeline&#239;&#188;&#140;&#229;&#190;&#136;&#228;&#184;&#141;&#231;&#187;&#153;&#229;&#138;&#155;","//         status.status_type = 'comments_timeline';","//       }","      ","//             status.created_at = new Date(data.timestamp * 1000);","//             status.timestamp = data.timestamp;","//             if(data.image){","//                 status.thumbnail_pic = data.image[0] + '/160';","//                 status.bmiddle_pic = data.image[0] + '/460';","//                 status.original_pic = data.image[0] + '/2000';","//             }","//       if (data.source) {","//         if(data.type == 4) { ","//           // &#229;&#155;&#158;&#229;&#164;&#141;","//           status.text = '@' + data.source.name + ' ' + status.text;","//           status.related_dialogue_url = 'http://t.qq.com/p/r/' + status.id;","//           status.in_reply_to_status_id = data.source.id;","//           status.in_reply_to_screen_name = data.source.nick;","//         } else {","//           status.retweeted_status = ","//             this.format_result_item(data.source, 'status', args, users);","//           // &#232;&#175;&#132;&#232;&#174;&#186;","//           if(play_load == 'comment') {","//             status.status = status.retweeted_status;","//             delete status.retweeted_status;","//           }","//         }","//       }","//       status.repost_count = data.count || 0;","//       status.comments_count = data.mcount || 0; // &#232;&#175;&#132;&#232;&#174;&#186;&#230;&#149;&#176;","//       status.source = data.from;","//       status.user = this.format_result_item(data, 'user', args, users);","//       // &#230;&#148;&#182;&#228;&#187;&#182;&#228;&#186;&#186;","// //      tohead: \"\"","// //      toisvip: 0","// //      toname: \"macgirl\"","// //      tonick: \"&#231;&#190;&#142;&#228;&#187;&#170;\"","//       if(data.toname) {","//         status.recipient = {","//           name: data.toname,","//           nick: data.tonick,","//           isvip: data.toisvip,","//           head: data.tohead","//         };","//         status.recipient = this.format_result_item(status.recipient, 'user', args, users);","//       }","      ","//       // &#229;&#166;&#130;&#230;&#158;&#156;&#230;&#156;&#137;text&#229;&#177;&#158;&#230;&#128;&#167;&#239;&#188;&#140;&#229;&#136;&#153;&#230;&#155;&#191;&#230;&#141;&#162;&#229;&#133;&#182;&#228;&#184;&#173;&#231;&#154;&#132;@xxx &#228;&#184;&#186; &#228;&#184;&#173;&#230;&#150;&#135;&#229;&#144;&#141;(@xxx)","//         if(status &amp;&amp; status.text) {","//           var matchs = status.text.match(this.ONLY_AT_USER_RE);","//           if(matchs) {","//             status.users = {};","//             for(var j=0; j&lt;matchs.length; j++) {","//               var name = matchs[j].trim().substring(1);","//               status.users[name] = users[name];","//             }","//           }","//         }","//         data = status;","};","","/**"," *","{ birth_day: 1,","  birth_month: 1,","  birth_year: 2010,","  city_code: '1',","  comp: null,","  country_code: '1',","  edu: null,","  email: '',","  exp: 56,","  fansnum: 3,","  favnum: 0,","  head: 'http://app.qlogo.cn/mbloghead/2045de7c75623f2c2b06',","  homecity_code: '',","  homecountry_code: '',","  homepage: '',","  homeprovince_code: '',","  hometown_code: '',","  idolnum: 46,","  industry_code: 0,","  introduction: '',","  isent: 0,","  ismyblack: 0,","  ismyfans: 0,","  ismyidol: 0,","  isrealname: 2,","  isvip: 0,","  level: 1,","  location: '&#228;&#184;&#173;&#229;&#155;&#189; &#230;&#157;&#173;&#229;&#183;&#158;',","  mutual_fans_num: 0,","  name: 'node-weibo',","  nick: 'node-weibo',","  openid: 'EA68676D5E9DA465822CD0CEB2DC6EF5',","  province_code: '33',","  regtime: 1348724066,","  send_private_flag: 2,","  sex: 1,","  tag: null,","  tweetinfo: ","   [ { city_code: '1',","       country_code: '1',","       emotiontype: 0,","       emotionurl: '',","       from: '&#232;&#133;&#190;&#232;&#174;&#175;&#229;&#190;&#174;&#229;&#141;&#154;',","       fromurl: 'http://t.qq.com\\n',","       geo: '',","       id: '70997003338788',","       image: null,","       latitude: '0',","       location: '&#228;&#184;&#173;&#229;&#155;&#189; &#230;&#157;&#173;&#229;&#183;&#158;',","       longitude: '0',","       music: null,","       origtext: '#&#230;&#150;&#176;&#228;&#186;&#186;&#230;&#138;&#165;&#229;&#136;&#176;# &#228;&#188;&#159;&#229;&#164;&#167;&#231;&#154;&#132;&#230;&#151;&#133;&#231;&#168;&#139;&#233;&#131;&#189;&#230;&#152;&#175;&#228;&#187;&#142;&#231;&#172;&#172;&#228;&#184;&#128;&#230;&#157;&#161;&#229;&#190;&#174;&#229;&#141;&#154;&#229;&#188;&#128;&#229;&#167;&#139;&#231;&#154;&#132;&#239;&#188;&#129;',","       province_code: '33',","       self: 1,","       status: 0,","       text: '#&#230;&#150;&#176;&#228;&#186;&#186;&#230;&#138;&#165;&#229;&#136;&#176;# &#228;&#188;&#159;&#229;&#164;&#167;&#231;&#154;&#132;&#230;&#151;&#133;&#231;&#168;&#139;&#233;&#131;&#189;&#230;&#152;&#175;&#228;&#187;&#142;&#231;&#172;&#172;&#228;&#184;&#128;&#230;&#157;&#161;&#229;&#190;&#174;&#229;&#141;&#154;&#229;&#188;&#128;&#229;&#167;&#139;&#231;&#154;&#132;&#239;&#188;&#129;',","       timestamp: 1348724111,","       type: 1,","       video: null } ],","  tweetnum: 1,","  verifyinfo: '' }"," */","TQQAPI.prototype.format_user = function (data, args) {","  var user = {};","  user.id = data.name;","  user.t_url = 'http://t.qq.com/' + data.name;","  user.screen_name = data.nick;","  user.name = data.name;","  user.location = data.location || '';","  user.description = data.introduction || '';","  // no url","  if (data.head) {","    user.profile_image_url = data.head + '/50'; // &#231;&#171;&#159;&#231;&#132;&#182;&#231;&#155;&#180;&#230;&#142;&#165;&#232;&#142;&#183;&#229;&#143;&#150;&#231;&#154;&#132;&#229;&#156;&#176;&#229;&#157;&#128;&#230;&#151;&#160;&#230;&#179;&#149;&#230;&#139;&#191;&#229;&#136;&#176;&#229;&#164;&#180;&#229;&#131;&#143;","    user.avatar_large = data.head + '/180';","  } else {","    user.profile_image_url = 'http://mat1.gtimg.com/www/mb/images/head_50.jpg';","    user.avatar_large = 'http://mat1.gtimg.com/www/mb/images/head_180.jpg';","  }","  user.gender = this.config.gender_map[data.sex||0];","  user.followers_count = data.fansnum || 0;","  user.friends_count = data.idolnum || 0;","  user.statuses_count = data.tweetnum || 0;","  user.favourites_count = data.favnum || 0;","  if (data.regtime) {","    user.created_at = new Date(data.regtime * 1000);","  }","  user.following = data.ismyidol || false;","  user.follow_me = data.ismyfans || false;","  // send_private_flag : &#230;&#152;&#175;&#229;&#144;&#166;&#229;&#133;&#129;&#232;&#174;&#184;&#230;&#137;&#128;&#230;&#156;&#137;&#228;&#186;&#186;&#231;&#187;&#153;&#229;&#189;&#147;&#229;&#137;&#141;&#231;&#148;&#168;&#230;&#136;&#183;&#229;&#143;&#145;&#231;&#167;&#129;&#228;&#191;&#161;&#239;&#188;&#140;0-&#228;&#187;&#133;&#230;&#156;&#137;&#229;&#129;&#182;&#229;&#131;&#143;&#239;&#188;&#140;1-&#229;&#144;&#141;&#228;&#186;&#186;+&#229;&#144;&#172;&#228;&#188;&#151;&#239;&#188;&#140;2-&#230;&#137;&#128;&#230;&#156;&#137;&#228;&#186;&#186;,","  user.allow_all_act_msg = data.send_private_flag === 2;","  // no geo_enabled","  user.verified = !!data.isvip;","  // no verified_type","  user.verified_reason = data.verifyinfo || '';","  // user.remark = ","  user.allow_all_comment = true;","  // user.online_status = true;","  user.bi_followers_count = data.mutual_fans_num || 0;","  // user.lang","  if (data.tweetinfo &amp;&amp; data.tweetinfo[0]) {","    user.status = this.format_status(data.tweetinfo[0], args);","  }","","  if (data.tag) {","    user.tags = data.tag;","  }","  return user;","};","","TQQAPI.prototype.format_count = function (count, args) {","  return count;","};","","TQQAPI.prototype.format_geo = function (data, args) {","  var geo = {","    longitude: data.longitude,","    latitude: data.latitude,","    // city_name string  City name \"&#229;&#185;&#191;&#229;&#183;&#158;\"","    // province_name string  Province name \"&#229;&#185;&#191;&#228;&#184;&#156;\"","    address: data.geo,","  };","  return geo;","};","","TQQAPI.prototype.format_comment = function (data, args) {","  var comment = this.format_status(data, args);","  if (comment.retweeted_status) {","    comment.status = comment.retweeted_status;","    delete comment.retweeted_status;","  }","  return comment;","};","","TQQAPI.prototype.format_message = function (message, args) {","  throw new Error('Must override this method.');","};","","TQQAPI.prototype.format_emotion = function (emotion, args) {","  throw new Error('Must override this method.');","};","","TQQAPI.prototype.format_favorite = function (status, args) {","  var favorite = {","    created_at: new Date(status.storetime * 1000),","    status: this.format_status(status)","  };","  return favorite;","};","","/**"," * Params converters"," */","","TQQAPI.prototype.convert_comment = function (comment) {","  // http://wiki.open.t.qq.com/index.php/%E5%BE%AE%E5%8D%9A%E7%9B%B8%E5%85%B3/%E7%82%B9%E8%AF%84%E4%B8%80%E6%9D%A1%E5%BE%AE%E5%8D%9A","  var data = {","    content: comment.comment,","    reid: comment.id","  };","  return data;","};","","TQQAPI.prototype.convert_status = function (status) {","  // syncflag &#229;&#190;&#174;&#229;&#141;&#154;&#229;&#144;&#140;&#230;&#173;&#165;&#229;&#136;&#176;&#231;&#169;&#186;&#233;&#151;&#180;&#229;&#136;&#134;&#228;&#186;&#171;&#230;&#160;&#135;&#232;&#174;&#176;&#239;&#188;&#136;&#229;&#143;&#175;&#233;&#128;&#137;&#239;&#188;&#140;0-&#229;&#144;&#140;&#230;&#173;&#165;&#239;&#188;&#140;1-&#228;&#184;&#141;&#229;&#144;&#140;&#230;&#173;&#165;&#239;&#188;&#140;&#233;&#187;&#152;&#232;&#174;&#164;&#228;&#184;&#186;0&#239;&#188;&#137;&#239;&#188;&#140;&#231;&#155;&#174;&#229;&#137;&#141;&#228;&#187;&#133;&#230;&#148;&#175;&#230;&#140;&#129;oauth1.0&#233;&#137;&#180;&#230;&#157;&#131;&#230;&#150;&#185;&#229;&#188;&#143;","  var data = {","    content: status.status","  };","  if (status.long) {","    data.longitude = status.long;","    data.latitude = status.lat;","  }","  if (status.id) {","    data.reid = status.id;","  }","  return data;","};","","TQQAPI.prototype.convert_user = function (user) {","  var data = {","    name: user.uid || user.screen_name","  };","  return data;","};","","TQQAPI.prototype.convert_ids = function (ids) {","  return {","    ids: ids,","    flag: '2'","  };","};","","/**"," * pageflag"," &#229;&#136;&#134;&#233;&#161;&#181;&#230;&#160;&#135;&#232;&#175;&#134;&#239;&#188;&#136;0&#239;&#188;&#154;&#231;&#172;&#172;&#228;&#184;&#128;&#233;&#161;&#181;&#239;&#188;&#140;1&#239;&#188;&#154;&#229;&#144;&#145;&#228;&#184;&#139;&#231;&#191;&#187;&#233;&#161;&#181;&#239;&#188;&#140;2&#239;&#188;&#154;&#229;&#144;&#145;&#228;&#184;&#138;&#231;&#191;&#187;&#233;&#161;&#181;&#239;&#188;&#137;"," pagetime"," &#230;&#156;&#172;&#233;&#161;&#181;&#232;&#181;&#183;&#229;&#167;&#139;&#230;&#151;&#182;&#233;&#151;&#180;&#239;&#188;&#136;&#231;&#172;&#172;&#228;&#184;&#128;&#233;&#161;&#181;&#239;&#188;&#154;&#229;&#161;&#171;0&#239;&#188;&#140;&#229;&#144;&#145;&#228;&#184;&#138;&#231;&#191;&#187;&#233;&#161;&#181;&#239;&#188;&#154;&#229;&#161;&#171;&#228;&#184;&#138;&#228;&#184;&#128;&#230;&#172;&#161;&#232;&#175;&#183;&#230;&#177;&#130;&#232;&#191;&#148;&#229;&#155;&#158;&#231;&#154;&#132;&#231;&#172;&#172;&#228;&#184;&#128;&#230;&#157;&#161;&#232;&#174;&#176;&#229;&#189;&#149;&#230;&#151;&#182;&#233;&#151;&#180;&#239;&#188;&#140;&#229;&#144;&#145;&#228;&#184;&#139;&#231;&#191;&#187;&#233;&#161;&#181;&#239;&#188;&#154;&#229;&#161;&#171;&#228;&#184;&#138;&#228;&#184;&#128;&#230;&#172;&#161;&#232;&#175;&#183;&#230;&#177;&#130;&#232;&#191;&#148;&#229;&#155;&#158;&#231;&#154;&#132;&#230;&#156;&#128;&#229;&#144;&#142;&#228;&#184;&#128;&#230;&#157;&#161;&#232;&#174;&#176;&#229;&#189;&#149;&#230;&#151;&#182;&#233;&#151;&#180;&#239;&#188;&#137;"," reqnum"," &#230;&#175;&#143;&#230;&#172;&#161;&#232;&#175;&#183;&#230;&#177;&#130;&#232;&#174;&#176;&#229;&#189;&#149;&#231;&#154;&#132;&#230;&#157;&#161;&#230;&#149;&#176;&#239;&#188;&#136;1-70&#230;&#157;&#161;&#239;&#188;&#137;"," type"," &#230;&#139;&#137;&#229;&#143;&#150;&#231;&#177;&#187;&#229;&#158;&#139;&#239;&#188;&#136;&#233;&#156;&#128;&#229;&#161;&#171;&#229;&#134;&#153;&#229;&#141;&#129;&#232;&#191;&#155;&#229;&#136;&#182;&#230;&#149;&#176;&#229;&#173;&#151;&#239;&#188;&#137;","0x1 &#229;&#142;&#159;&#229;&#136;&#155;&#229;&#143;&#145;&#232;&#161;&#168; 0x2 &#232;&#189;&#172;&#232;&#189;&#189; &#229;&#166;&#130;&#233;&#156;&#128;&#230;&#139;&#137;&#229;&#143;&#150;&#229;&#164;&#154;&#228;&#184;&#170;&#231;&#177;&#187;&#229;&#158;&#139;&#232;&#175;&#183;&#228;&#189;&#191;&#231;&#148;&#168;|&#239;&#188;&#140;&#229;&#166;&#130;(0x1|0x2)&#229;&#190;&#151;&#229;&#136;&#176;3&#239;&#188;&#140;&#229;&#136;&#153;type=3&#229;&#141;&#179;&#229;&#143;&#175;&#239;&#188;&#140;&#229;&#161;&#171;&#233;&#155;&#182;&#232;&#161;&#168;&#231;&#164;&#186;&#230;&#139;&#137;&#229;&#143;&#150;&#230;&#137;&#128;&#230;&#156;&#137;&#231;&#177;&#187;&#229;&#158;&#139; "," contenttype"," &#229;&#134;&#133;&#229;&#174;&#185;&#232;&#191;&#135;&#230;&#187;&#164;&#227;&#128;&#130;0-&#232;&#161;&#168;&#231;&#164;&#186;&#230;&#137;&#128;&#230;&#156;&#137;&#231;&#177;&#187;&#229;&#158;&#139;&#239;&#188;&#140;1-&#229;&#184;&#166;&#230;&#150;&#135;&#230;&#156;&#172;&#239;&#188;&#140;2-&#229;&#184;&#166;&#233;&#147;&#190;&#230;&#142;&#165;&#239;&#188;&#140;4-&#229;&#184;&#166;&#229;&#155;&#190;&#231;&#137;&#135;&#239;&#188;&#140;8-&#229;&#184;&#166;&#232;&#167;&#134;&#233;&#162;&#145;&#239;&#188;&#140;0x10-&#229;&#184;&#166;&#233;&#159;&#179;&#233;&#162;&#145;","  &#229;&#187;&#186;&#232;&#174;&#174;&#228;&#184;&#141;&#228;&#189;&#191;&#231;&#148;&#168;contenttype&#228;&#184;&#186;1&#231;&#154;&#132;&#231;&#177;&#187;&#229;&#158;&#139;&#239;&#188;&#140;&#229;&#166;&#130;&#230;&#158;&#156;&#232;&#166;&#129;&#230;&#139;&#137;&#229;&#143;&#150;&#229;&#143;&#170;&#230;&#156;&#137;&#230;&#150;&#135;&#230;&#156;&#172;&#231;&#154;&#132;&#229;&#190;&#174;&#229;&#141;&#154;&#239;&#188;&#140;&#229;&#187;&#186;&#232;&#174;&#174;&#228;&#189;&#191;&#231;&#148;&#168;0x80"," * "," */","TQQAPI.prototype.convert_cursor = function (cursor) {","  var data = {};","  // type: &#230;&#139;&#137;&#229;&#143;&#150;&#231;&#177;&#187;&#229;&#158;&#139;, 0x1 &#229;&#142;&#159;&#229;&#136;&#155;&#229;&#143;&#145;&#232;&#161;&#168; 0x2 &#232;&#189;&#172;&#232;&#189;&#189; 0x8 &#229;&#155;&#158;&#229;&#164;&#141; 0x10 &#231;&#169;&#186;&#229;&#155;&#158; 0x20 &#230;&#143;&#144;&#229;&#143;&#138; 0x40 &#231;&#130;&#185;&#232;&#175;&#132;","  data.type = String(0x1 | 0x2 | 0x8 | 0x10 | 0x20);","  data.contenttype = '0';","  data.reqnum = cursor.count;","  if (cursor.max_id) {","    // get older statuses","    data.pageflag = '1';","    data.pagetime = cursor.max_time;","    data.lastid = cursor.max_id;","  } else if (cursor.since_id) {","    // get newer statuses","    // 0&#239;&#188;&#154;&#231;&#172;&#172;&#228;&#184;&#128;&#233;&#161;&#181;&#239;&#188;&#140;1&#239;&#188;&#154;&#229;&#144;&#145;&#228;&#184;&#139;&#231;&#191;&#187;&#233;&#161;&#181;&#239;&#188;&#140;2&#239;&#188;&#154;&#229;&#144;&#145;&#228;&#184;&#138;&#231;&#191;&#187;&#233;&#161;&#181;","    data.pageflag = '2';","    data.pagetime = cursor.since_time;","    // data.lastid = cursor.sina_id;","  } else {","    // top page","    data.pageflag = '0';","    data.pagetime = '0';","    data.lastid = '0';","  }","  if (typeof cursor.callback === 'function') {","    data = cursor.callback(data);","  }","  return data;","};","","/**"," * Status"," */","","TQQAPI.prototype.repost_timeline = function (user, cursor, callback) {","  cursor.callback = function (data) {","    data.rootid = cursor.id;","    data.flag = '0';","    // twitterid &#229;&#190;&#174;&#229;&#141;&#154;id&#239;&#188;&#140;&#228;&#184;&#142;pageflag&#227;&#128;&#129;pagetime&#229;&#133;&#177;&#229;&#144;&#140;&#228;&#189;&#191;&#231;&#148;&#168;&#239;&#188;&#140;&#229;&#174;&#158;&#231;&#142;&#176;&#231;&#191;&#187;&#233;&#161;&#181;&#229;&#138;&#159;&#232;&#131;&#189;&#239;&#188;&#136;&#231;&#172;&#172;1&#233;&#161;&#181;&#229;&#161;&#171;0&#239;&#188;&#140;&#231;&#187;&#167;&#231;&#187;&#173;&#229;&#144;&#145;&#228;&#184;&#139;&#231;&#191;&#187;&#233;&#161;&#181;&#239;&#188;&#140;&#229;&#161;&#171;&#228;&#184;&#138;&#228;&#184;&#128;&#230;&#172;&#161;&#232;&#175;&#183;&#230;&#177;&#130;&#232;&#191;&#148;&#229;&#155;&#158;&#231;&#154;&#132;&#230;&#156;&#128;&#229;&#144;&#142;&#228;&#184;&#128;&#230;&#157;&#161;&#232;&#174;&#176;&#229;&#189;&#149;id&#239;&#188;&#137;","    if (data.lastid) {","      data.twitterid = data.lastid;","      delete data.lastid;","    }","    return data;","  };","  return TQQAPI.super_.prototype.repost_timeline.call(this, user, cursor, callback);","};","","TQQAPI.prototype.user_timeline = function (user, cursor, callback) {","  cursor.callback = function (data) {","    if (cursor.uid || cursor.screen_name) {","      data.name = cursor.uid || cursor.screen_name;","    }","    return data;","  };","  return TQQAPI.super_.prototype.user_timeline.call(this, user, cursor, callback);","};","","TQQAPI.prototype.search = function (user, query, cursor, callback) {","  cursor = cursor || {};","  var q = {","    keyword: query.q","  };","  if (query.long &amp;&amp; query.lat &amp;&amp; query.radius) {","    q.longitude = query.long;","    q.latitude = query.lat;","    q.radius = query.radius;","  }","  cursor.callback = function (data) {","    data.pagesize = data.reqnum || 20;","    return data;","  };","  return TQQAPI.super_.prototype.search.call(this, user, q, cursor, callback);","};","","/**"," * Comment"," */","","TQQAPI.prototype.comments_timeline = function (user, cursor, callback) {","  cursor.callback = function (data) {","    data.type = String(0x40);","    return data;","  };","  return TQQAPI.super_.prototype.comments_timeline.call(this, user, cursor, callback);","};","","TQQAPI.prototype.comments = function (user, cursor, callback) {","  cursor.callback = function (data) {","    data.rootid = cursor.id;","    data.flag = '1';","    if (data.lastid) {","      data.twitterid = data.lastid;","      delete data.lastid;","    }","    return data;","  };","  return TQQAPI.super_.prototype.comments.call(this, user, cursor, callback);","};","","TQQAPI.prototype.comment_destroy = function (user, cid, callback) {","  callback(new TypeError('comment_destroy not support.'));","};","","/**"," * Favorite"," */","","TQQAPI.prototype.favorite_show = function (user, id, callback) {","  var self = this;","  self.show(user, id, function (err, status) {","    if (err) {","      err.name = self.errorname('favorite_show');","      return callback(err);","    }","    var favorite = {","      status: status,","      created_at: status.created_at","    };","    callback(null, favorite);","  });","};",""];
