// weibo/lib-cov/weibo.js
/* automatically generated by JSCoverage - do not edit */
if (typeof _$jscoverage === 'undefined') _$jscoverage = {};
if (! _$jscoverage['weibo.js']) {
  _$jscoverage['weibo.js'] = [];
  _$jscoverage['weibo.js'][7] = 0;
  _$jscoverage['weibo.js'][13] = 0;
  _$jscoverage['weibo.js'][14] = 0;
  _$jscoverage['weibo.js'][15] = 0;
  _$jscoverage['weibo.js'][16] = 0;
  _$jscoverage['weibo.js'][19] = 0;
  _$jscoverage['weibo.js'][20] = 0;
  _$jscoverage['weibo.js'][22] = 0;
  _$jscoverage['weibo.js'][40] = 0;
  _$jscoverage['weibo.js'][43] = 0;
  _$jscoverage['weibo.js'][44] = 0;
  _$jscoverage['weibo.js'][50] = 0;
  _$jscoverage['weibo.js'][51] = 0;
  _$jscoverage['weibo.js'][59] = 0;
  _$jscoverage['weibo.js'][60] = 0;
  _$jscoverage['weibo.js'][63] = 0;
  _$jscoverage['weibo.js'][64] = 0;
  _$jscoverage['weibo.js'][65] = 0;
  _$jscoverage['weibo.js'][66] = 0;
  _$jscoverage['weibo.js'][67] = 0;
  _$jscoverage['weibo.js'][68] = 0;
  _$jscoverage['weibo.js'][72] = 0;
  _$jscoverage['weibo.js'][73] = 0;
  _$jscoverage['weibo.js'][77] = 0;
  _$jscoverage['weibo.js'][80] = 0;
  _$jscoverage['weibo.js'][81] = 0;
  _$jscoverage['weibo.js'][82] = 0;
  _$jscoverage['weibo.js'][83] = 0;
  _$jscoverage['weibo.js'][87] = 0;
  _$jscoverage['weibo.js'][90] = 0;
  _$jscoverage['weibo.js'][91] = 0;
  _$jscoverage['weibo.js'][92] = 0;
  _$jscoverage['weibo.js'][93] = 0;
  _$jscoverage['weibo.js'][94] = 0;
  _$jscoverage['weibo.js'][95] = 0;
  _$jscoverage['weibo.js'][96] = 0;
  _$jscoverage['weibo.js'][97] = 0;
  _$jscoverage['weibo.js'][100] = 0;
  _$jscoverage['weibo.js'][103] = 0;
  _$jscoverage['weibo.js'][104] = 0;
  _$jscoverage['weibo.js'][105] = 0;
  _$jscoverage['weibo.js'][106] = 0;
  _$jscoverage['weibo.js'][107] = 0;
  _$jscoverage['weibo.js'][109] = 0;
  _$jscoverage['weibo.js'][110] = 0;
  _$jscoverage['weibo.js'][112] = 0;
  _$jscoverage['weibo.js'][113] = 0;
  _$jscoverage['weibo.js'][115] = 0;
  _$jscoverage['weibo.js'][118] = 0;
  _$jscoverage['weibo.js'][119] = 0;
  _$jscoverage['weibo.js'][122] = 0;
  _$jscoverage['weibo.js'][123] = 0;
  _$jscoverage['weibo.js'][126] = 0;
  _$jscoverage['weibo.js'][127] = 0;
  _$jscoverage['weibo.js'][128] = 0;
  _$jscoverage['weibo.js'][131] = 0;
  _$jscoverage['weibo.js'][132] = 0;
  _$jscoverage['weibo.js'][133] = 0;
  _$jscoverage['weibo.js'][134] = 0;
  _$jscoverage['weibo.js'][135] = 0;
  _$jscoverage['weibo.js'][142] = 0;
  _$jscoverage['weibo.js'][143] = 0;
  _$jscoverage['weibo.js'][144] = 0;
}
_$jscoverage['weibo.js'][7]++;
"use strict";
_$jscoverage['weibo.js'][13]++;
var TBaseOauthV2 = require("./tbase_oauth_v2");
_$jscoverage['weibo.js'][14]++;
var inherits = require("util").inherits;
_$jscoverage['weibo.js'][15]++;
var utils = require("./utils");
_$jscoverage['weibo.js'][16]++;
var weiboutil = require("./weibo_util");
_$jscoverage['weibo.js'][19]++;
function WeiboAPI(options) {
  _$jscoverage['weibo.js'][20]++;
  WeiboAPI.super_.call(this);
  _$jscoverage['weibo.js'][22]++;
  var config = utils.extend({}, options, {host: "https://api.weibo.com/2", user_home_url: "http://weibo.com/n/", search_url: "http://s.weibo.com/weibo/", oauth_host: "https://api.weibo.com/oauth2", oauth_authorize: "/authorize", oauth_access_token: "/access_token", verify_credentials: "/users/show", comments: "/comments/show", comment_create: "/comments/create", comment_reply: "/comments/reply", comment_destroy: "/comments/destroy", support_search: false});
  _$jscoverage['weibo.js'][40]++;
  this.init(config);
}
_$jscoverage['weibo.js'][43]++;
inherits(WeiboAPI, TBaseOauthV2);
_$jscoverage['weibo.js'][44]++;
module.exports = WeiboAPI;
_$jscoverage['weibo.js'][50]++;
WeiboAPI.prototype.get_result_items = (function (data, playload, args) {
  _$jscoverage['weibo.js'][51]++;
  return data.statuses || data.comments || data.reposts || data.messages || data.favorites || data;
});
_$jscoverage['weibo.js'][59]++;
WeiboAPI.prototype.format_search_status = (function (status, args) {
  _$jscoverage['weibo.js'][60]++;
  return status;
});
_$jscoverage['weibo.js'][63]++;
WeiboAPI.prototype.format_status = (function (status, args) {
  _$jscoverage['weibo.js'][64]++;
  status.id = status.idstr;
  _$jscoverage['weibo.js'][65]++;
  status.created_at = new Date(status.created_at);
  _$jscoverage['weibo.js'][66]++;
  if (status.user) {
    _$jscoverage['weibo.js'][67]++;
    status.user = this.format_user(status.user, args);
    _$jscoverage['weibo.js'][68]++;
    status.t_url = "http://weibo.com/" + status.user.id + "/" + weiboutil.mid2url(status.mid);
  }
  _$jscoverage['weibo.js'][72]++;
  if (status.geo && status.geo.type === "Point" && status.geo.coordinates) {
    _$jscoverage['weibo.js'][73]++;
    var geo = {latitude: String(status.geo.coordinates[0]), longitude: String(status.geo.coordinates[1])};
    _$jscoverage['weibo.js'][77]++;
    status.geo = geo;
  }
  _$jscoverage['weibo.js'][80]++;
  if (status.retweeted_status) {
    _$jscoverage['weibo.js'][81]++;
    status.retweeted_status = this.format_status(status.retweeted_status, args);
    _$jscoverage['weibo.js'][82]++;
    if (! status.retweeted_status.t_url) {
      _$jscoverage['weibo.js'][83]++;
      status.retweeted_status.t_url = "http://weibo.com/" + status.user.id + "/" + weiboutil.mid2url(status.retweeted_status.mid);
    }
  }
  _$jscoverage['weibo.js'][87]++;
  return status;
});
_$jscoverage['weibo.js'][90]++;
WeiboAPI.prototype.format_user = (function (user, args) {
  _$jscoverage['weibo.js'][91]++;
  user.id = user.idstr;
  _$jscoverage['weibo.js'][92]++;
  user.created_at = new Date(user.created_at);
  _$jscoverage['weibo.js'][93]++;
  user.t_url = "http://weibo.com/" + (user.domain || user.id);
  _$jscoverage['weibo.js'][94]++;
  if (user.status) {
    _$jscoverage['weibo.js'][95]++;
    user.status = this.format_status(user.status, args);
    _$jscoverage['weibo.js'][96]++;
    if (! user.status.t_url) {
      _$jscoverage['weibo.js'][97]++;
      user.status.t_url = user.t_url + "/" + weiboutil.mid2url(user.status.mid || user.status.id);
    }
  }
  _$jscoverage['weibo.js'][100]++;
  return user;
});
_$jscoverage['weibo.js'][103]++;
WeiboAPI.prototype.format_comment = (function (comment, args) {
  _$jscoverage['weibo.js'][104]++;
  comment.id = comment.idstr;
  _$jscoverage['weibo.js'][105]++;
  comment.created_at = new Date(comment.created_at);
  _$jscoverage['weibo.js'][106]++;
  if (comment.user) {
    _$jscoverage['weibo.js'][107]++;
    comment.user = this.format_user(comment.user, args);
  }
  _$jscoverage['weibo.js'][109]++;
  if (comment.status) {
    _$jscoverage['weibo.js'][110]++;
    comment.status = this.format_status(comment.status, args);
  }
  _$jscoverage['weibo.js'][112]++;
  if (comment.reply_comment) {
    _$jscoverage['weibo.js'][113]++;
    comment.reply_comment = this.format_comment(comment.reply_comment, args);
  }
  _$jscoverage['weibo.js'][115]++;
  return comment;
});
_$jscoverage['weibo.js'][118]++;
WeiboAPI.prototype.format_message = (function (message, args) {
  _$jscoverage['weibo.js'][119]++;
  return message;
});
_$jscoverage['weibo.js'][122]++;
WeiboAPI.prototype.format_emotion = (function (emotion, args) {
  _$jscoverage['weibo.js'][123]++;
  return emotion;
});
_$jscoverage['weibo.js'][126]++;
WeiboAPI.prototype.format_count = (function (count, args) {
  _$jscoverage['weibo.js'][127]++;
  count.id = String(count.id);
  _$jscoverage['weibo.js'][128]++;
  return count;
});
_$jscoverage['weibo.js'][131]++;
WeiboAPI.prototype.format_favorite = (function (favorite, args) {
  _$jscoverage['weibo.js'][132]++;
  favorite.status = this.format_status(favorite.status);
  _$jscoverage['weibo.js'][133]++;
  favorite.created_at = new Date(favorite.favorited_time);
  _$jscoverage['weibo.js'][134]++;
  delete favorite.favorited_time;
  _$jscoverage['weibo.js'][135]++;
  return favorite;
});
_$jscoverage['weibo.js'][142]++;
WeiboAPI.prototype.verify_credentials = (function (user, callback) {
  _$jscoverage['weibo.js'][143]++;
  var uid = user.uid || user.id;
  _$jscoverage['weibo.js'][144]++;
  return this.user_show(user, uid, null, callback);
});
_$jscoverage['weibo.js'].source = ["/*!"," * node-weibo - lib/weibo.js"," * Copyright(c) 2012 fengmk2 &lt;fengmk2@gmail.com&gt;"," * MIT Licensed"," */","","\"use strict\";","","/**"," * Module dependencies."," */","","var TBaseOauthV2 = require('./tbase_oauth_v2');","var inherits = require('util').inherits;","var utils = require('./utils');","var weiboutil = require('./weibo_util');","","","function WeiboAPI(options) {","  WeiboAPI.super_.call(this);","","  var config = utils.extend({}, options, {","    host:                 'https://api.weibo.com/2',","    user_home_url:        'http://weibo.com/n/',","    search_url:           'http://s.weibo.com/weibo/',","","    oauth_host:           'https://api.weibo.com/oauth2',","    oauth_authorize:      '/authorize',","    oauth_access_token:   '/access_token',","    verify_credentials:   '/users/show',","","    comments:             '/comments/show',","    comment_create:       '/comments/create',","    comment_reply:        '/comments/reply',","    comment_destroy:      '/comments/destroy',","","    support_search: false,","  });","","  this.init(config);","}","","inherits(WeiboAPI, TBaseOauthV2);","module.exports = WeiboAPI;","","/**"," * Result getters"," */","","WeiboAPI.prototype.get_result_items = function (data, playload, args) {","  return data.statuses || data.comments || data.reposts ||","    data.messages || data.favorites || data;","};","","/**"," * Result formatters"," */","","WeiboAPI.prototype.format_search_status = function (status, args) {","  return status;","};","","WeiboAPI.prototype.format_status = function (status, args) {","  status.id = status.idstr;","  status.created_at = new Date(status.created_at);","  if (status.user) {","    status.user = this.format_user(status.user, args);","    status.t_url = 'http://weibo.com/' + status.user.id + '/' + weiboutil.mid2url(status.mid);","  }","","  // geo: { type: 'Point', coordinates: [ 22.354231, 113.421234 ] } latitude, longitude","  if (status.geo &amp;&amp; status.geo.type === 'Point' &amp;&amp; status.geo.coordinates) {","    var geo = {","      latitude: String(status.geo.coordinates[0]),","      longitude: String(status.geo.coordinates[1]),","    };","    status.geo = geo;","  }","","  if (status.retweeted_status) {","    status.retweeted_status = this.format_status(status.retweeted_status, args);","    if (!status.retweeted_status.t_url) {","      status.retweeted_status.t_url =","        'http://weibo.com/' + status.user.id + '/' + weiboutil.mid2url(status.retweeted_status.mid);","    }","  }","  return status;","};","","WeiboAPI.prototype.format_user = function (user, args) {","  user.id = user.idstr;","  user.created_at = new Date(user.created_at);","  user.t_url = 'http://weibo.com/' + (user.domain || user.id);","  if (user.status) {","    user.status = this.format_status(user.status, args);","    if (!user.status.t_url) {","      user.status.t_url = user.t_url + '/' + weiboutil.mid2url(user.status.mid || user.status.id);","    }","  }","  return user;","};","","WeiboAPI.prototype.format_comment = function (comment, args) {","  comment.id = comment.idstr;","  comment.created_at = new Date(comment.created_at);","  if (comment.user) {","    comment.user = this.format_user(comment.user, args);","  }","  if (comment.status) {","    comment.status = this.format_status(comment.status, args);","  }","  if (comment.reply_comment) {","    comment.reply_comment = this.format_comment(comment.reply_comment, args);","  }","  return comment;","};","","WeiboAPI.prototype.format_message = function (message, args) {","  return message;","};","","WeiboAPI.prototype.format_emotion = function (emotion, args) {","  return emotion;","};","","WeiboAPI.prototype.format_count = function (count, args) {","  count.id = String(count.id);","  return count;","};","","WeiboAPI.prototype.format_favorite = function (favorite, args) {","  favorite.status = this.format_status(favorite.status);","  favorite.created_at = new Date(favorite.favorited_time);","  delete favorite.favorited_time;","  return favorite;","};","","/**"," * User"," */","","WeiboAPI.prototype.verify_credentials = function (user, callback) {","  var uid = user.uid || user.id;","  return this.user_show(user, uid, null, callback);","};",""];
